// @tanstack/intent@0.0.44/dist/artifact-coverage-DgWuVqUp.mjs
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { parse } from "yaml";

//#region src/artifact-coverage.ts
function isRecord(value) {
	return !!value && typeof value === "object" && !Array.isArray(value);
}
function stringValue(value) {
	return typeof value === "string" && value.trim() ? value : void 0;
}
function stringArray(value) {
	if (!Array.isArray(value)) return [];
	return value.filter((entry) => typeof entry === "string");
}
function detectArtifactKind(fileName) {
	if (fileName.endsWith("skill_tree.yaml")) return "skill-tree";
	if (fileName.endsWith("domain_map.yaml")) return "domain-map";
	return null;
}
function readArtifactYaml(artifactPath, warnings) {
	let parsed;
	try {
		parsed = parse(readFileSync(artifactPath, "utf8"));
	} catch (err) {
		warnings.push({
			artifactPath,
			message: `Invalid YAML: ${err instanceof Error ? err.message : String(err)}`
		});
		return null;
	}
	if (!isRecord(parsed)) {
		warnings.push({
			artifactPath,
			message: "Artifact YAML must contain an object at the top level"
		});
		return null;
	}
	return parsed;
}
function parseArtifactFile(artifactPath, kind, parsed) {
	const library = isRecord(parsed.library) ? parsed.library : {};
	return {
		path: artifactPath,
		kind,
		libraryName: stringValue(library.name),
		libraryVersion: stringValue(library.version)
	};
}
function parseSkills(artifactPath, kind, parsed) {
	if (!Array.isArray(parsed.skills)) return [];
	const skills = [];
	for (const skill of parsed.skills) {
		if (!isRecord(skill)) continue;
		const packagePath = stringValue(skill.package);
		const packages = stringArray(skill.packages);
		if (packagePath) packages.push(packagePath);
		skills.push({
			artifactPath,
			artifactKind: kind,
			name: stringValue(skill.name),
			slug: stringValue(skill.slug),
			path: stringValue(skill.path),
			package: packagePath,
			packages: [...new Set(packages)].sort((a, b) => a.localeCompare(b)),
			sources: stringArray(skill.sources),
			covers: stringArray(skill.covers)
		});
	}
	return skills;
}
function parseCoverageIgnores(artifactPath, parsed) {
	const ignored = (isRecord(parsed.coverage) ? parsed.coverage : null)?.ignored_packages;
	if (!Array.isArray(ignored)) return [];
	const ignores = [];
	for (const entry of ignored) {
		if (typeof entry === "string") {
			if (entry.trim()) ignores.push({
				packageName: entry,
				artifactPath
			});
			continue;
		}
		if (!isRecord(entry)) continue;
		const packageName = stringValue(entry.name);
		if (!packageName) continue;
		ignores.push({
			packageName,
			reason: stringValue(entry.reason),
			artifactPath
		});
	}
	return ignores;
}
function readIntentArtifacts(root) {
	const artifactsDir = join(root, "_artifacts");
	if (!existsSync(artifactsDir)) return null;
	const warnings = [];
	const skillTrees = [];
	const domainMaps = [];
	const skills = [];
	const ignoredPackages = [];
	for (const entry of readdirSync(artifactsDir, { withFileTypes: true })) {
		if (!entry.isFile()) continue;
		const kind = detectArtifactKind(entry.name);
		if (!kind) continue;
		const artifactPath = join(artifactsDir, entry.name);
		const parsed = readArtifactYaml(artifactPath, warnings);
		if (!parsed) continue;
		const artifactFile = parseArtifactFile(artifactPath, kind, parsed);
		if (kind === "skill-tree") skillTrees.push(artifactFile);
		else domainMaps.push(artifactFile);
		skills.push(...parseSkills(artifactPath, kind, parsed));
		ignoredPackages.push(...parseCoverageIgnores(artifactPath, parsed));
	}
	return {
		root,
		artifactsDir,
		skillTrees: skillTrees.sort((a, b) => a.path.localeCompare(b.path)),
		domainMaps: domainMaps.sort((a, b) => a.path.localeCompare(b.path)),
		skills: skills.sort((a, b) => `${a.artifactPath}:${a.slug ?? a.name ?? ""}`.localeCompare(`${b.artifactPath}:${b.slug ?? b.name ?? ""}`)),
		ignoredPackages: ignoredPackages.sort((a, b) => a.packageName.localeCompare(b.packageName)),
		warnings
	};
}

//#endregion
export { readIntentArtifacts as t };
