// @tanstack/intent@0.0.44/dist/cli-support-BNDR3l1g.mjs
import { t as resolveProjectContext } from "./project-context-BRIMzhi4.mjs";
import { t as fail } from "./cli-error-BrMXlbtx.mjs";
import { existsSync, readFileSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";

//#region src/cli-output.ts
function printWarnings(warnings) {
	if (warnings.length === 0) return;
	console.log("Warnings:");
	for (const warning of warnings) console.log(`  ⚠ ${warning}`);
}

//#endregion
//#region src/cli-support.ts
const INTENT_CHECK_SKILLS_WORKFLOW_VERSION = 3;
function getMetaDir() {
	return join(dirname(fileURLToPath(import.meta.url)), "..", "meta");
}
function getCheckSkillsWorkflowAdvisories(root) {
	const workflowPath = join(root, ".github", "workflows", "check-skills.yml");
	if (!existsSync(workflowPath)) return [];
	let content;
	try {
		content = readFileSync(workflowPath, "utf8");
	} catch {
		return [];
	}
	const versionMatch = content.match(/intent-workflow-version:\s*(\d+)/);
	if ((versionMatch ? Number(versionMatch[1]) : 0) >= INTENT_CHECK_SKILLS_WORKFLOW_VERSION) return [];
	return [`Intent workflow update available: run \`npx @tanstack/intent@latest setup\` to refresh ${relative(process.cwd(), workflowPath) || workflowPath}.`];
}
async function scanIntentsOrFail(options) {
	const { scanForIntents } = await import("./scanner-BV7X5n_z.mjs");
	try {
		return scanForIntents(void 0, options);
	} catch (err) {
		fail(err instanceof Error ? err.message : String(err));
	}
}
function scanOptionsFromGlobalFlags(options) {
	if (options.global && options.globalOnly) fail("Use either --global or --global-only, not both.");
	if (options.globalOnly) return { scope: "global" };
	if (options.global) return { scope: "local-and-global" };
	return { scope: "local" };
}
function coreOptionsFromGlobalFlags(options) {
	if (options.global && options.globalOnly) fail("Use either --global or --global-only, not both.");
	return {
		debug: options.debug,
		exclude: Array.isArray(options.exclude) ? options.exclude : options.exclude ? [options.exclude] : void 0,
		global: options.global,
		globalOnly: options.globalOnly
	};
}
function formatDebugValue(value) {
	if (Array.isArray(value)) return value.length > 0 ? value.join(", ") : "(none)";
	return String(value);
}
function printDebugInfo(title, fields) {
	console.error(`Debug: ${title}`);
	for (const [label, value] of fields) console.error(`  ${label}: ${formatDebugValue(value)}`);
}
async function resolveStaleTargets(targetDir) {
	const resolvedRoot = targetDir ? resolve(process.cwd(), targetDir) : process.cwd();
	const context = resolveProjectContext({
		cwd: process.cwd(),
		targetPath: targetDir
	});
	const workflowAdvisories = getCheckSkillsWorkflowAdvisories(context.workspaceRoot ?? context.packageRoot ?? resolvedRoot);
	const { buildWorkspaceCoverageSignals, checkStaleness, readPackageName } = await import("./staleness-CB3qpVCq.mjs");
	const isWorkspaceRootTarget = context.workspaceRoot !== null && resolvedRoot === context.workspaceRoot;
	if (context.packageRoot && !isWorkspaceRootTarget && (context.targetSkillsDir !== null || context.workspaceRoot === null)) return {
		reports: [await checkStaleness(context.packageRoot, readPackageName(context.packageRoot), context.workspaceRoot ?? context.packageRoot)],
		workflowAdvisories
	};
	const { findWorkspaceRoot, getWorkspaceInfo } = await import("./workspace-patterns-Bjq_cXww.mjs");
	const workspaceRoot = findWorkspaceRoot(resolvedRoot);
	const workspaceInfo = workspaceRoot ? getWorkspaceInfo(workspaceRoot) : null;
	if (workspaceInfo) {
		const reports = await Promise.all(workspaceInfo.packageDirsWithSkills.map((packageDir) => checkStaleness(packageDir, readPackageName(packageDir), workspaceInfo.root)));
		const { readIntentArtifacts } = await import("./artifact-coverage-DA26utB1.mjs");
		const artifacts = existsSync(join(workspaceInfo.root, "_artifacts")) ? readIntentArtifacts(workspaceInfo.root) : null;
		const coverageSignals = buildWorkspaceCoverageSignals({
			artifactRoot: workspaceInfo.root,
			artifacts,
			packageDirs: workspaceInfo.packageDirs
		});
		if (coverageSignals.length > 0) reports.push({
			library: relative(process.cwd(), workspaceInfo.root) || "workspace",
			currentVersion: null,
			skillVersion: null,
			versionDrift: null,
			skills: [],
			signals: coverageSignals
		});
		if (reports.length > 0) return {
			reports,
			workflowAdvisories
		};
	}
	if (existsSync(join(resolvedRoot, "skills"))) return {
		reports: [await checkStaleness(resolvedRoot, readPackageName(resolvedRoot))],
		workflowAdvisories
	};
	const staleResult = await scanIntentsOrFail();
	return {
		reports: await Promise.all(staleResult.packages.map((pkg) => checkStaleness(pkg.packageRoot, pkg.name))),
		workflowAdvisories
	};
}

//#endregion
export { printDebugInfo as a, scanOptionsFromGlobalFlags as c, getMetaDir as i, printWarnings as l, coreOptionsFromGlobalFlags as n, resolveStaleTargets as o, getCheckSkillsWorkflowAdvisories as r, scanIntentsOrFail as s, INTENT_CHECK_SKILLS_WORKFLOW_VERSION as t };
