// @tanstack/intent@0.0.44/dist/core-qlBI_aQL.mjs
import { i as getDeps, l as resolveDepDir, u as toPosixPath } from "./utils-9fhWAVua.mjs";
import { n as scanIntentPackageAtRoot, r as createIntentFsCache, t as scanForIntents } from "./scanner-DaG82zNc.mjs";
import { n as findWorkspacePackages } from "./workspace-patterns-BffPlZ1D.mjs";
import { i as parseSkillUse, n as formatSkillUse } from "./skill-use-umYvZl94.mjs";
import { t as resolveProjectContext } from "./project-context-BRIMzhi4.mjs";
import { a as compileExcludePatterns, c as warningMentionsPackage, i as resolveSkillUse, o as getEffectiveExcludePatterns, r as resolveSkillEntry, s as isPackageExcluded, t as ResolveSkillUseError } from "./resolver-D9cjsz11.mjs";
import { existsSync } from "node:fs";
import { dirname, isAbsolute, join, relative, resolve } from "node:path";

//#region src/core/markdown.ts
function resolveFromCwd$1(path) {
	return resolve(process.cwd(), path);
}
function splitDestinationSuffix(destination) {
	const hashIndex = destination.indexOf("#");
	const queryIndex = destination.indexOf("?");
	const suffixIndex = hashIndex === -1 ? queryIndex : queryIndex === -1 ? hashIndex : Math.min(hashIndex, queryIndex);
	if (suffixIndex === -1) return {
		pathPart: destination,
		suffix: ""
	};
	return {
		pathPart: destination.slice(0, suffixIndex),
		suffix: destination.slice(suffixIndex)
	};
}
function isExternalOrAbsoluteDestination(destination) {
	return destination === "" || destination.startsWith("#") || destination.startsWith("?") || destination.startsWith("//") || /^[A-Za-z][A-Za-z0-9+.-]*:/.test(destination) || isAbsolute(destination);
}
function findClosingBracket(line, start) {
	let depth = 0;
	for (let index = start; index < line.length; index++) {
		const char = line[index];
		if (char === "\\") {
			index++;
			continue;
		}
		if (char === "[") {
			depth++;
			continue;
		}
		if (char === "]") {
			depth--;
			if (depth === 0) return index;
		}
	}
	return -1;
}
function findClosingParen(line, start) {
	for (let index = start; index < line.length; index++) {
		const char = line[index];
		if (char === "\\") {
			index++;
			continue;
		}
		if (char === ")") return index;
	}
	return -1;
}
function readBareDestination(line, start) {
	let depth = 0;
	for (let index = start; index < line.length; index++) {
		const char = line[index];
		if (char === "\\") {
			index++;
			continue;
		}
		if (char === "(") {
			depth++;
			continue;
		}
		if (char === ")") {
			if (depth === 0) return {
				destinationEnd: index,
				endParen: index
			};
			depth--;
			continue;
		}
		if (/\s/.test(char) && depth === 0) {
			const endParen = findClosingParen(line, index);
			if (endParen === -1) return null;
			return {
				destinationEnd: index,
				endParen
			};
		}
	}
	return null;
}
function readMarkdownDestination(line, start) {
	let cursor = start;
	while (cursor < line.length && /\s/.test(line[cursor])) cursor++;
	if (line[cursor] === "<") {
		const destinationStart = cursor + 1;
		const destinationEnd = line.indexOf(">", destinationStart);
		if (destinationEnd === -1) return null;
		const endParen = findClosingParen(line, destinationEnd + 1);
		if (endParen === -1) return null;
		return {
			destination: line.slice(destinationStart, destinationEnd),
			destinationStart,
			destinationEnd,
			endParen
		};
	}
	const read = readBareDestination(line, cursor);
	if (!read) return null;
	return {
		destination: line.slice(cursor, read.destinationEnd),
		destinationStart: cursor,
		destinationEnd: read.destinationEnd,
		endParen: read.endParen
	};
}
function getCodeFenceMarker(line) {
	const marker = line.match(/^\s*(`{3,}|~{3,})/)?.[1]?.[0];
	return marker === "`" || marker === "~" ? marker : null;
}
function rewriteMarkdownDestination({ context, destination }) {
	if (isExternalOrAbsoluteDestination(destination)) return destination;
	const { pathPart, suffix } = splitDestinationSuffix(destination);
	if (isExternalOrAbsoluteDestination(pathPart)) return destination;
	const resolvedDestinationPath = resolve(context.skillDir, pathPart);
	const relativeToPackageRoot = relative(context.resolvedPackageRoot, resolvedDestinationPath);
	if (relativeToPackageRoot.startsWith("..") || isAbsolute(relativeToPackageRoot)) return destination;
	const relativeToCwd = relative(context.cwd, resolvedDestinationPath);
	return `${toPosixPath(relativeToCwd && !relativeToCwd.startsWith("..") && !isAbsolute(relativeToCwd) ? relativeToCwd : resolvedDestinationPath)}${suffix}`;
}
function rewriteMarkdownLineDestinations({ context, line }) {
	if (!line.includes("[")) return line;
	let output = "";
	let cursor = 0;
	while (cursor < line.length) {
		const nextCodeStart = line.indexOf("`", cursor);
		const nextLinkStart = line.indexOf("[", cursor);
		if (nextLinkStart === -1) {
			output += line.slice(cursor);
			break;
		}
		if (nextCodeStart !== -1 && nextCodeStart < nextLinkStart) {
			output += line.slice(cursor, nextCodeStart);
			cursor = nextCodeStart;
			const codeStart = cursor;
			while (cursor < line.length && line[cursor] === "`") cursor++;
			const marker = line.slice(codeStart, cursor);
			const codeEnd = line.indexOf(marker, cursor);
			if (codeEnd === -1) {
				output += line.slice(codeStart);
				break;
			}
			output += line.slice(codeStart, codeEnd + marker.length);
			cursor = codeEnd + marker.length;
			continue;
		}
		const linkStart = nextLinkStart > 0 && line[nextLinkStart - 1] === "!" ? nextLinkStart - 1 : nextLinkStart;
		output += line.slice(cursor, linkStart);
		const labelEnd = findClosingBracket(line, nextLinkStart);
		if (labelEnd === -1) {
			output += line.slice(linkStart);
			break;
		}
		if (line[labelEnd + 1] !== "(") {
			output += line.slice(linkStart, nextLinkStart + 1);
			cursor = nextLinkStart + 1;
			continue;
		}
		const destination = readMarkdownDestination(line, labelEnd + 2);
		if (!destination) {
			output += line.slice(linkStart, nextLinkStart + 1);
			cursor = nextLinkStart + 1;
			continue;
		}
		const rewritten = rewriteMarkdownDestination({
			context,
			destination: destination.destination
		});
		output += line.slice(linkStart, destination.destinationStart) + rewritten + line.slice(destination.destinationEnd, destination.endParen + 1);
		cursor = destination.endParen + 1;
	}
	return output;
}
function rewriteLoadedSkillMarkdownDestinations({ content, cwd, packageRoot, skillFilePath }) {
	const context = {
		cwd,
		resolvedPackageRoot: resolveFromCwd$1(packageRoot),
		skillDir: dirname(skillFilePath)
	};
	let inFence = null;
	const parts = content.split(/(\r?\n)/);
	let output = "";
	for (let index = 0; index < parts.length; index += 2) {
		const line = parts[index] ?? "";
		const newline = parts[index + 1] ?? "";
		const marker = getCodeFenceMarker(line);
		if (inFence) {
			output += line + newline;
			if (marker === inFence) inFence = null;
			continue;
		}
		if (marker) {
			inFence = marker;
			output += line + newline;
			continue;
		}
		output += rewriteMarkdownLineDestinations({
			context,
			line
		}) + newline;
	}
	return output;
}

//#endregion
//#region src/core/load-resolution.ts
function readWorkspacePackageInfos(context, fsCache) {
	const dirs = /* @__PURE__ */ new Set();
	if (context.packageRoot) dirs.add(context.packageRoot);
	if (context.workspaceRoot) {
		dirs.add(context.workspaceRoot);
		for (const dir of findWorkspacePackages(context.workspaceRoot)) dirs.add(dir);
	}
	return [...dirs].flatMap((dir) => {
		const packageJson = fsCache.readPackageJson(dir);
		if (!packageJson) return [];
		return [{
			dir,
			name: typeof packageJson.name === "string" ? packageJson.name : null,
			packageJson
		}];
	});
}
function addCandidateDir(candidates, seen, dir) {
	if (!dir) return;
	const key = resolve(dir);
	if (seen.has(key)) return;
	seen.add(key);
	candidates.push(dir);
}
function findVisibleDependencyDir(packageName, fromDir) {
	let dir = fromDir;
	let prev;
	while (dir !== prev) {
		const candidate = join(dir, "node_modules", packageName);
		if (existsSync(join(candidate, "package.json"))) return candidate;
		prev = dir;
		dir = dirname(dir);
	}
	return null;
}
function resolveDependencyPackageDir(packageName, fromDir) {
	return findVisibleDependencyDir(packageName, fromDir) ?? resolveDepDir(packageName, fromDir);
}
function workspacePackageDeclaresDependency(packageJson, packageName) {
	return getDeps(packageJson).includes(packageName);
}
function hasYarnPnpFile(dir) {
	return dir !== null && (existsSync(join(dir, ".pnp.cjs")) || existsSync(join(dir, ".pnp.js")));
}
function shouldSkipFastPathForYarnPnp(context, cwd) {
	return hasYarnPnpFile(cwd) || hasYarnPnpFile(context.workspaceRoot);
}
function getDirectLoadFastPathCandidateDirs(packageName, context, cwd) {
	const candidates = [];
	const seen = /* @__PURE__ */ new Set();
	if (!context.workspaceRoot) {
		addCandidateDir(candidates, seen, resolveDependencyPackageDir(packageName, context.packageRoot ?? cwd));
		return candidates;
	}
	addCandidateDir(candidates, seen, resolveDependencyPackageDir(packageName, context.packageRoot ?? context.workspaceRoot));
	if (context.workspaceRoot && context.workspaceRoot !== context.packageRoot) addCandidateDir(candidates, seen, resolveDependencyPackageDir(packageName, context.workspaceRoot));
	return candidates;
}
function getWorkspaceLoadFastPathCandidateDirs(packageName, context, fsCache) {
	const candidates = [];
	const seen = /* @__PURE__ */ new Set();
	const workspacePackages = readWorkspacePackageInfos(context, fsCache);
	for (const pkg of workspacePackages) if (pkg.name === packageName) addCandidateDir(candidates, seen, pkg.dir);
	for (const pkg of workspacePackages) {
		if (!workspacePackageDeclaresDependency(pkg.packageJson, packageName)) continue;
		addCandidateDir(candidates, seen, resolveDependencyPackageDir(packageName, pkg.dir));
	}
	return candidates;
}
function resolveScannedPackageSkill(scanned, parsedUse) {
	const pkg = scanned.package;
	if (!pkg || pkg.name !== parsedUse.packageName) return null;
	const skill = resolveSkillEntry(pkg.name, parsedUse.skillName, pkg.skills).skill;
	if (!skill) return null;
	return {
		packageName: pkg.name,
		skillName: skill.name,
		path: skill.path,
		source: pkg.source,
		version: pkg.version,
		packageRoot: pkg.packageRoot,
		warnings: scanned.warnings.filter((warning) => warningMentionsPackage(warning, pkg.name)),
		conflict: null
	};
}
function resolveFromPackageRoots(packageRoots, parsedUse, cwd, fsCache) {
	for (const packageRoot of packageRoots) {
		const scanned = scanIntentPackageAtRoot(packageRoot, {
			fallbackName: parsedUse.packageName,
			fsCache,
			projectRoot: cwd,
			skillNameHint: parsedUse.skillName
		});
		const directResolved = resolveScannedPackageSkill(scanned, parsedUse);
		if (directResolved) return directResolved;
		if (scanned.package?.name === parsedUse.packageName) {
			const fallbackResolved = resolveScannedPackageSkill(scanIntentPackageAtRoot(packageRoot, {
				fallbackName: parsedUse.packageName,
				fsCache,
				projectRoot: cwd
			}), parsedUse);
			if (fallbackResolved) return fallbackResolved;
		}
	}
	return null;
}
function resolveSkillUseFastPath(parsedUse, options, context = resolveProjectContext({ cwd: process.cwd() }), cwd = context.cwd, fsCache = createIntentFsCache()) {
	if (options.globalOnly) return null;
	if (shouldSkipFastPathForYarnPnp(context, cwd)) return null;
	const directResolved = resolveFromPackageRoots(getDirectLoadFastPathCandidateDirs(parsedUse.packageName, context, cwd), parsedUse, cwd, fsCache);
	if (directResolved) return directResolved;
	if (!context.workspaceRoot) return null;
	return resolveFromPackageRoots(getWorkspaceLoadFastPathCandidateDirs(parsedUse.packageName, context, fsCache), parsedUse, cwd, fsCache);
}

//#endregion
//#region src/core.ts
var IntentCoreError = class extends Error {
	constructor(code, message, options = {}) {
		super(message);
		this.name = "IntentCoreError";
		this.code = code;
		if (options.suggestedSkills) this.suggestedSkills = options.suggestedSkills;
	}
};
function toScanOptions(options) {
	if (options.global && options.globalOnly) throw new IntentCoreError("invalid-options", "Use either global or globalOnly, not both.");
	if (options.globalOnly) return { scope: "global" };
	if (options.global) return { scope: "local-and-global" };
	return { scope: "local" };
}
function getScanScope(options) {
	return options.scope ?? (options.includeGlobal ? "local-and-global" : "local");
}
function withFsCache(options, fsCache) {
	return {
		...options,
		fsCache
	};
}
function resolveCoreCwd(options) {
	return resolve(process.cwd(), options.cwd ?? process.cwd());
}
function listIntentSkills(options = {}) {
	const cwd = resolveCoreCwd(options);
	const scanOptions = toScanOptions(options);
	const fsCache = createIntentFsCache();
	const projectContext = resolveProjectContext({ cwd });
	const scanResult = scanForIntents(cwd, withFsCache(scanOptions, fsCache));
	const excludePatterns = getEffectiveExcludePatterns(options, projectContext);
	const excludeMatchers = compileExcludePatterns(excludePatterns);
	const excludedPackages = scanResult.packages.filter((pkg) => isPackageExcluded(pkg.name, excludeMatchers)).map((pkg) => pkg.name);
	const packages = scanResult.packages.filter((pkg) => !isPackageExcluded(pkg.name, excludeMatchers));
	const skills = packages.flatMap((pkg) => pkg.skills.map((skill) => {
		return {
			use: formatSkillUse(pkg.name, skill.name),
			packageName: pkg.name,
			packageRoot: pkg.packageRoot,
			packageVersion: pkg.version,
			packageSource: pkg.source,
			skillName: skill.name,
			description: skill.description,
			type: skill.type,
			framework: skill.framework
		};
	}));
	const result = {
		packageManager: scanResult.packageManager,
		skills,
		packages: packages.map((pkg) => ({
			name: pkg.name,
			version: pkg.version,
			source: pkg.source,
			packageRoot: pkg.packageRoot,
			skillCount: pkg.skills.length
		})),
		warnings: scanResult.warnings.filter((warning) => !excludedPackages.some((packageName) => warningMentionsPackage(warning, packageName))),
		conflicts: scanResult.conflicts.filter((conflict) => !isPackageExcluded(conflict.packageName, excludeMatchers))
	};
	if (options.debug) result.debug = {
		cwd,
		scope: getScanScope(scanOptions),
		excludes: excludePatterns,
		packageCount: result.packages.length,
		skillCount: result.skills.length,
		warningCount: result.warnings.length,
		conflictCount: result.conflicts.length,
		scan: scanResult.stats ?? fsCache.getStats()
	};
	return result;
}
function resolveFromCwd(cwd, path) {
	return resolve(cwd, path);
}
function isResolvedPathInsidePackageRoot(path, packageRoot) {
	const relativePath = relative(packageRoot, path);
	return relativePath === "" || !relativePath.startsWith("..") && !isAbsolute(relativePath);
}
function toResolvedIntentSkill(cwd, use, resolved, readFs, debug) {
	let realResolvedPath;
	try {
		realResolvedPath = readFs.realpathSync.native(resolveFromCwd(cwd, resolved.path));
	} catch {
		throw new IntentCoreError("skill-file-not-found", `Resolved skill file was not found: ${resolved.path}`);
	}
	const realPackageRoot = readFs.realpathSync.native(resolveFromCwd(cwd, resolved.packageRoot));
	if (!isResolvedPathInsidePackageRoot(realResolvedPath, realPackageRoot)) throw new IntentCoreError("skill-path-outside-package", `Resolved skill path for "${use}" is outside package root: ${resolved.path}`);
	const result = {
		path: resolved.path,
		packageRoot: resolved.packageRoot,
		packageName: resolved.packageName,
		skillName: resolved.skillName,
		version: resolved.version,
		source: resolved.source,
		warnings: resolved.warnings,
		conflict: resolved.conflict
	};
	if (debug) result.debug = debug;
	return {
		realPackageRoot,
		realResolvedPath,
		readFs,
		result
	};
}
function createLoadedSkillDebug({ cwd, excludes, scan, resolution, resolved, scope }) {
	return {
		cwd,
		scope,
		resolution,
		excludes,
		packageName: resolved.packageName,
		skillName: resolved.skillName,
		version: resolved.version,
		source: resolved.source,
		path: resolved.path,
		warningCount: resolved.warnings.length,
		scan
	};
}
function resolveIntentSkillInCwd(cwd, use, options = {}) {
	let parsedUse;
	try {
		parsedUse = parseSkillUse(use);
	} catch (err) {
		throw new IntentCoreError("invalid-skill-use", err instanceof Error ? err.message : String(err));
	}
	const fsCache = createIntentFsCache();
	const projectContext = resolveProjectContext({ cwd });
	const excludePatterns = getEffectiveExcludePatterns(options, projectContext);
	const excludeMatchers = compileExcludePatterns(excludePatterns);
	if (isPackageExcluded(parsedUse.packageName, excludeMatchers)) throw new IntentCoreError("package-excluded", `Cannot load skill use "${use}": package "${parsedUse.packageName}" is excluded by Intent configuration.`);
	const scanOptions = toScanOptions(options);
	const scope = getScanScope(scanOptions);
	const fastPathResolved = resolveSkillUseFastPath(parsedUse, options, projectContext, cwd, fsCache);
	if (fastPathResolved) return toResolvedIntentSkill(cwd, use, fastPathResolved, fsCache.getReadFs(), options.debug ? createLoadedSkillDebug({
		cwd,
		excludes: excludePatterns,
		resolution: "fast-path",
		resolved: fastPathResolved,
		scan: fsCache.getStats(),
		scope
	}) : void 0);
	const scanResult = scanForIntents(cwd, withFsCache(scanOptions, fsCache));
	let resolved;
	try {
		resolved = resolveSkillUse(use, scanResult);
	} catch (err) {
		if (err instanceof ResolveSkillUseError) throw new IntentCoreError(err.code, err.message, { suggestedSkills: err.suggestedSkills });
		throw err;
	}
	return toResolvedIntentSkill(cwd, use, resolved, fsCache.getReadFs(), options.debug ? createLoadedSkillDebug({
		cwd,
		excludes: excludePatterns,
		resolution: "full-scan",
		resolved,
		scan: scanResult.stats ?? fsCache.getStats(),
		scope
	}) : void 0);
}
function resolveIntentSkill(use, options = {}) {
	return resolveIntentSkillInCwd(resolveCoreCwd(options), use, options).result;
}
function loadIntentSkill(use, options = {}) {
	const cwd = resolveCoreCwd(options);
	const resolved = resolveIntentSkillInCwd(cwd, use, options);
	const content = rewriteLoadedSkillMarkdownDestinations({
		content: resolved.readFs.readFileSync(resolved.realResolvedPath, "utf8"),
		cwd,
		packageRoot: resolved.realPackageRoot,
		skillFilePath: resolved.realResolvedPath
	});
	return {
		...resolved.result,
		content
	};
}

//#endregion
export { resolveIntentSkill as i, listIntentSkills as n, loadIntentSkill as r, IntentCoreError as t };
