// @tanstack/intent@0.0.44/dist/display-0FnzlKRq.mjs
import "./utils-9fhWAVua.mjs";
import { n as isStableLoadPath, t as formatRuntimeSkillLookupHint } from "./skill-paths-B-j0dWDA.mjs";

//#region src/display.ts
function padColumn(text, width) {
	return text.length >= width ? text + "  " : text.padEnd(width);
}
function printTable(headers, rows) {
	const widths = headers.map((h, i) => Math.max(h.length, ...rows.map((r) => (r[i] ?? "").length)) + 2);
	const headerLine = headers.map((h, i) => padColumn(h, widths[i])).join("");
	const separator = widths.map((w) => "─".repeat(w)).join("");
	console.log(headerLine);
	console.log(separator);
	for (const row of rows) console.log(row.map((cell, i) => padColumn(cell, widths[i])).join(""));
}
function printSkillLine(displayName, skill, indent, opts) {
	const nameStr = " ".repeat(indent) + displayName;
	const padding = " ".repeat(Math.max(2, opts.nameWidth - nameStr.length));
	const typeCol = opts.showTypes ? (skill.type ? `[${skill.type}]` : "").padEnd(14) : "";
	console.log(`${nameStr}${padding}${typeCol}${skill.description}`);
	if (skill.loadCommand) console.log(`${" ".repeat(indent + 2)}Load: ${skill.loadCommand}`);
	if (skill.path) {
		const pathIndent = " ".repeat(indent + 2);
		if (isStableLoadPath(skill.path)) console.log(`${pathIndent}${skill.path}`);
		else console.log(`${pathIndent}${formatRuntimeSkillLookupHint({
			packageName: opts.packageName,
			skillName: skill.name
		})}`);
	}
}
function printSkillTree(skills, opts) {
	const roots = [];
	const children = /* @__PURE__ */ new Map();
	const printedSkills = /* @__PURE__ */ new Set();
	for (const skill of skills) {
		const slashIdx = skill.name.indexOf("/");
		if (slashIdx === -1) roots.push(skill.name);
		else {
			const parent = skill.name.slice(0, slashIdx);
			if (!children.has(parent)) children.set(parent, []);
			children.get(parent).push(skill);
		}
	}
	if (roots.length === 0) {
		for (const skill of skills) if (!roots.includes(skill.name)) roots.push(skill.name);
	}
	for (const rootName of roots) {
		const rootSkill = skills.find((s) => s.name === rootName);
		if (!rootSkill) continue;
		printSkillLine(rootName, rootSkill, 4, opts);
		printedSkills.add(rootSkill.name);
		for (const sub of children.get(rootName) ?? []) {
			printSkillLine(sub.name.slice(sub.name.indexOf("/") + 1), sub, 6, opts);
			printedSkills.add(sub.name);
		}
	}
	for (const skill of skills) {
		if (printedSkills.has(skill.name)) continue;
		printSkillLine(skill.name, skill, 4, opts);
	}
}
function computeSkillNameWidth(allPackageSkills) {
	let max = 0;
	for (const skills of allPackageSkills) for (const s of skills) {
		const slashIdx = s.name.indexOf("/");
		const displayName = slashIdx === -1 ? s.name : s.name.slice(slashIdx + 1);
		const indent = slashIdx === -1 ? 4 : 6;
		max = Math.max(max, indent + displayName.length);
	}
	return max + 2;
}

//#endregion
export { computeSkillNameWidth, printSkillTree, printTable };
