// @tanstack/intent@0.0.44/dist/index.mjs
import { c as parseFrontmatter, i as getDeps, l as resolveDepDir, r as findSkillFiles } from "./utils-9fhWAVua.mjs";
import "./skill-paths-B-j0dWDA.mjs";
import { t as scanForIntents } from "./scanner-DaG82zNc.mjs";
import "./workspace-patterns-BffPlZ1D.mjs";
import { t as readIntentArtifacts } from "./artifact-coverage-DgWuVqUp.mjs";
import { n as checkStaleness } from "./staleness-BGFfic-Y.mjs";
import { n as collectStaleReviewItems, r as createFailedStaleReviewItem, t as buildStaleReviewBody } from "./workflow-review-D9Fg1G0P.mjs";
import { i as parseSkillUse, n as formatSkillUse, r as isSkillUseParseError, t as SkillUseParseError } from "./skill-use-umYvZl94.mjs";
import "./project-context-BRIMzhi4.mjs";
import { i as resolveSkillUse, n as isResolveSkillUseError, t as ResolveSkillUseError } from "./resolver-D9cjsz11.mjs";
import { r as runSetupGithubActions, t as runEditPackageJson } from "./setup-BklJkH_6.mjs";
import { readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { execFileSync, execSync } from "node:child_process";

//#region src/feedback.ts
const META_FEEDBACK_REPO = "TanStack/intent";
const SECRET_PATTERNS = [
	/(?:ghp|gho|ghu|ghs|ghr)_[A-Za-z0-9_]{36,}/,
	/(?:sk|pk)[-_](?:live|test)[-_][A-Za-z0-9]{24,}/,
	/AKIA[0-9A-Z]{16}/,
	/-----BEGIN (?:RSA |EC )?PRIVATE KEY-----/,
	/eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}/,
	/(?:Bearer|token)\s+[A-Za-z0-9_\-.~+/]{20,}/i,
	/[A-Za-z0-9]{32,}(?=.*(?:key|secret|token|password))/i
];
function containsSecrets(text) {
	return SECRET_PATTERNS.some((pattern) => pattern.test(text));
}
function hasGhCli() {
	try {
		execSync("gh --version", { stdio: "ignore" });
		return true;
	} catch {
		return false;
	}
}
function getHomeConfigDir() {
	return process.env.XDG_CONFIG_HOME ?? join(process.env.HOME ?? process.env.USERPROFILE ?? "", ".config");
}
function parseFrequency(value) {
	if (value === "always" || value === "never") return value;
	if (typeof value !== "string") return null;
	const match = /^every-(\d+)$/.exec(value);
	if (!match) return null;
	const count = Number(match[1]);
	return Number.isInteger(count) && count > 0 ? `every-${count}` : null;
}
function readFrequency(filePath) {
	try {
		return parseFrequency(JSON.parse(readFileSync(filePath, "utf8")).feedback?.frequency);
	} catch {
		return null;
	}
}
function resolveFrequency(root) {
	const userFrequency = readFrequency(join(getHomeConfigDir(), "intent", "config.json"));
	if (userFrequency) return userFrequency;
	const projectFrequency = readFrequency(join(root, "intent.config.json"));
	if (projectFrequency) return projectFrequency;
	return "every-5";
}
const REQUIRED_FIELDS = [
	"skill",
	"package",
	"skillVersion",
	"task",
	"whatWorked",
	"whatFailed",
	"missing",
	"selfCorrections",
	"userRating"
];
function validatePayload(payload) {
	const errors = [];
	if (!payload || typeof payload !== "object") return {
		valid: false,
		errors: ["Payload must be a JSON object"]
	};
	const obj = payload;
	for (const field of REQUIRED_FIELDS) if (typeof obj[field] !== "string" || obj[field].trim() === "") errors.push(`Missing or empty required field: ${field}`);
	if (obj.userRating && ![
		"good",
		"mixed",
		"bad"
	].includes(obj.userRating)) errors.push("userRating must be one of: good, mixed, bad");
	if (containsSecrets(Object.values(obj).filter((v) => typeof v === "string").join("\n"))) errors.push("Payload appears to contain secrets or tokens — submission rejected");
	return {
		valid: errors.length === 0,
		errors
	};
}
const META_REQUIRED_FIELDS = [
	"metaSkill",
	"library",
	"agentUsed",
	"artifactQuality",
	"whatWorked",
	"whatFailed",
	"suggestions",
	"userRating"
];
const VALID_META_SKILLS = [
	"domain-discovery",
	"tree-generator",
	"generate-skill",
	"skill-staleness-check"
];
const VALID_AGENTS = [
	"claude-code",
	"cursor",
	"copilot",
	"codex",
	"other"
];
const VALID_QUALITY_RATINGS = [
	"good",
	"mixed",
	"bad"
];
const VALID_INTERVIEW_QUALITY_RATINGS = [
	"good",
	"mixed",
	"bad",
	"skipped"
];
const VALID_FAILURE_MODE_QUALITY_RATINGS = [
	"good",
	"mixed",
	"bad",
	"not-applicable"
];
function validateMetaPayload(payload) {
	const errors = [];
	if (!payload || typeof payload !== "object") return {
		valid: false,
		errors: ["Payload must be a JSON object"]
	};
	const obj = payload;
	for (const field of META_REQUIRED_FIELDS) if (typeof obj[field] !== "string" || obj[field].trim() === "") errors.push(`Missing or empty required field: ${field}`);
	if (obj.metaSkill && !VALID_META_SKILLS.includes(obj.metaSkill)) errors.push(`metaSkill must be one of: ${VALID_META_SKILLS.join(", ")}`);
	if (obj.agentUsed && !VALID_AGENTS.includes(obj.agentUsed)) errors.push(`agentUsed must be one of: ${VALID_AGENTS.join(", ")}`);
	if (obj.artifactQuality && !VALID_QUALITY_RATINGS.includes(obj.artifactQuality)) errors.push("artifactQuality must be one of: good, mixed, bad");
	if (obj.userRating && !VALID_QUALITY_RATINGS.includes(obj.userRating)) errors.push("userRating must be one of: good, mixed, bad");
	if (obj.interviewQuality && !VALID_INTERVIEW_QUALITY_RATINGS.includes(obj.interviewQuality)) errors.push("interviewQuality must be one of: good, mixed, bad, skipped");
	if (obj.failureModeQuality && !VALID_FAILURE_MODE_QUALITY_RATINGS.includes(obj.failureModeQuality)) errors.push("failureModeQuality must be one of: good, mixed, bad, not-applicable");
	if (containsSecrets(Object.values(obj).filter((v) => typeof v === "string").join("\n"))) errors.push("Payload appears to contain secrets or tokens — submission rejected");
	return {
		valid: errors.length === 0,
		errors
	};
}
function metaToMarkdown(payload) {
	const lines = [
		`# Meta-Skill Feedback: ${payload.metaSkill}`,
		"",
		`**Library:** ${payload.library}`,
		`**Agent:** ${payload.agentUsed}`,
		`**Artifact quality:** ${payload.artifactQuality}`,
		`**Rating:** ${payload.userRating}`
	];
	if (payload.interviewQuality) lines.push(`**Interview quality:** ${payload.interviewQuality}`);
	if (payload.failureModeQuality) lines.push(`**Failure mode quality:** ${payload.failureModeQuality}`);
	lines.push("", "## What Worked", payload.whatWorked, "", "## What Failed", payload.whatFailed, "", "## Suggestions", payload.suggestions);
	return lines.join("\n") + "\n";
}
function toMarkdown(payload) {
	const lines = [
		`# Skill Feedback: ${payload.skill}`,
		"",
		`**Package:** ${payload.package}`,
		`**Skill version:** ${payload.skillVersion}`,
		`**Rating:** ${payload.userRating}`,
		"",
		"## Task",
		payload.task,
		"",
		"## What Worked",
		payload.whatWorked,
		"",
		"## What Failed",
		payload.whatFailed,
		"",
		"## Missing",
		payload.missing,
		"",
		"## Self-Corrections",
		payload.selfCorrections
	];
	if (payload.userComments) lines.push("", "## User Comments", payload.userComments);
	return lines.join("\n") + "\n";
}
function submitFeedback(payload, repo, opts) {
	const md = toMarkdown(payload);
	if (opts.ghAvailable) try {
		execFileSync("gh", [
			"issue",
			"create",
			"--repo",
			repo,
			"--title",
			`Skill Feedback: ${payload.skill} (${payload.userRating})`,
			"--body",
			"-"
		], {
			input: md,
			stdio: [
				"pipe",
				"pipe",
				"pipe"
			]
		});
		return {
			method: "gh",
			detail: `Submitted issue to ${repo}`
		};
	} catch (err) {
		const msg = err instanceof Error ? err.message : String(err);
		console.error(`GitHub submission failed: ${msg}`);
		console.error("Falling back to file output.");
	}
	if (opts.outputPath) {
		writeFileSync(opts.outputPath, md, "utf8");
		return {
			method: "file",
			detail: `Saved to ${opts.outputPath}`
		};
	}
	return {
		method: "stdout",
		detail: md
	};
}
function submitMetaFeedback(payload, opts) {
	const md = metaToMarkdown(payload);
	if (opts.ghAvailable) try {
		execFileSync("gh", [
			"issue",
			"create",
			"--repo",
			META_FEEDBACK_REPO,
			"--title",
			`Meta-Skill Feedback: ${payload.metaSkill} (${payload.userRating})`,
			"--label",
			`skill:${payload.metaSkill}`,
			"--body",
			"-"
		], {
			input: md,
			stdio: [
				"pipe",
				"pipe",
				"pipe"
			]
		});
		return {
			method: "gh",
			detail: `Submitted issue to ${META_FEEDBACK_REPO}`
		};
	} catch (err) {
		const msg = err instanceof Error ? err.message : String(err);
		console.error(`GitHub submission failed: ${msg}`);
		console.error("Falling back to file output.");
	}
	if (opts.outputPath) {
		writeFileSync(opts.outputPath, md, "utf8");
		return {
			method: "file",
			detail: `Saved to ${opts.outputPath}`
		};
	}
	return {
		method: "stdout",
		detail: md
	};
}

//#endregion
export { ResolveSkillUseError, SkillUseParseError, buildStaleReviewBody, checkStaleness, collectStaleReviewItems, containsSecrets, createFailedStaleReviewItem, findSkillFiles, formatSkillUse, getDeps, hasGhCli, isResolveSkillUseError, isSkillUseParseError, metaToMarkdown, parseFrontmatter, parseSkillUse, readIntentArtifacts, resolveDepDir, resolveFrequency, resolveSkillUse, runEditPackageJson, runSetupGithubActions, scanForIntents, submitFeedback, submitMetaFeedback, toMarkdown, validateMetaPayload, validatePayload };
