// @tanstack/intent@0.0.44/dist/list-DrChOziX.mjs
import "./utils-9fhWAVua.mjs";
import "./skill-paths-B-j0dWDA.mjs";
import "./scanner-DaG82zNc.mjs";
import "./workspace-patterns-BffPlZ1D.mjs";
import "./project-context-BRIMzhi4.mjs";
import "./resolver-D9cjsz11.mjs";
import { n as listIntentSkills } from "./core-qlBI_aQL.mjs";
import { a as printDebugInfo, l as printWarnings, n as coreOptionsFromGlobalFlags } from "./cli-support-BNDR3l1g.mjs";
import { t as formatIntentCommand } from "./command-runner-C0yCOHLi.mjs";

//#region src/commands/list.ts
function printListDebug(result) {
	if (!result.debug) return;
	printDebugInfo("intent list", [
		["cwd", result.debug.cwd],
		["scope", result.debug.scope],
		["excludes", result.debug.excludes],
		["packages", result.debug.packageCount],
		["skills", result.debug.skillCount],
		["warnings", result.debug.warningCount],
		["conflicts", result.debug.conflictCount],
		["packageJsonReadCount", result.debug.scan.packageJsonReadCount],
		["packageJsonCacheHits", result.debug.scan.packageJsonCacheHits]
	]);
}
function printVersionConflicts(result) {
	if (result.conflicts.length === 0) return;
	console.log("\nVersion conflicts:\n");
	for (const conflict of result.conflicts) {
		console.log(`  ${conflict.packageName} -> using ${conflict.chosen.version}`);
		console.log(`    chosen: ${conflict.chosen.packageRoot}`);
		for (const variant of conflict.variants) {
			if (variant.packageRoot === conflict.chosen.packageRoot) continue;
			console.log(`    also found: ${variant.version} at ${variant.packageRoot}`);
		}
		console.log();
	}
}
function groupSkillsByPackageRoot(skills) {
	const grouped = /* @__PURE__ */ new Map();
	for (const skill of skills) {
		const packageSkills = grouped.get(skill.packageRoot);
		if (packageSkills) packageSkills.push(skill);
		else grouped.set(skill.packageRoot, [skill]);
	}
	return grouped;
}
function getPackageSkills(pkg, skillsByPackageRoot) {
	return skillsByPackageRoot.get(pkg.packageRoot) ?? [];
}
function formatLoadCommand(skill, packageManager, scopeFlag) {
	return formatIntentCommand(packageManager, `load ${skill.use}${scopeFlag}`);
}
async function runListCommand(options, _scanIntentsOrFail) {
	const result = listIntentSkills(coreOptionsFromGlobalFlags(options));
	printListDebug(result);
	if (options.json) {
		const { debug: _debug, packageManager: _packageManager, ...jsonResult } = result;
		console.log(JSON.stringify(jsonResult, null, 2));
		return;
	}
	const { computeSkillNameWidth, printSkillTree, printTable } = await import("./display-0FnzlKRq.mjs");
	if (result.packages.length === 0) {
		console.log("No intent-enabled packages found.");
		if (result.warnings.length > 0) {
			console.log();
			printWarnings(result.warnings);
		}
		return;
	}
	console.log(`\n${result.packages.length} intent-enabled packages, ${result.skills.length} skills\n`);
	printTable([
		"PACKAGE",
		"SOURCE",
		"VERSION",
		"SKILLS"
	], result.packages.map((pkg) => [
		pkg.name,
		pkg.source,
		pkg.version,
		String(pkg.skillCount)
	]));
	printVersionConflicts(result);
	const skillsByPackageRoot = groupSkillsByPackageRoot(result.skills);
	const nameWidth = computeSkillNameWidth(result.packages.map((pkg) => getPackageSkills(pkg, skillsByPackageRoot).map((skill) => ({
		name: skill.skillName,
		description: skill.description,
		type: skill.type
	}))));
	const showTypes = result.skills.some((skill) => skill.type);
	const scopeFlag = options.globalOnly ? " --global-only" : options.global ? " --global" : "";
	console.log(`\nSkills:\n`);
	for (const pkg of result.packages) {
		console.log(`  ${pkg.name}`);
		printSkillTree(getPackageSkills(pkg, skillsByPackageRoot).map((skill) => ({
			name: skill.skillName,
			description: skill.description,
			loadCommand: formatLoadCommand(skill, result.packageManager, scopeFlag),
			type: skill.type
		})), {
			nameWidth,
			packageName: pkg.name,
			showTypes
		});
		console.log();
	}
	console.log("Feedback:");
	console.log("  Submit feedback on skill usage to help maintainers improve the skills.");
	console.log("  Load: node_modules/@tanstack/intent/meta/feedback-collection/SKILL.md");
	console.log();
	printWarnings(result.warnings);
}

//#endregion
export { runListCommand };
