// @tanstack/intent@0.0.44/dist/load-ll7s5E3J.mjs
import "./utils-9fhWAVua.mjs";
import "./skill-paths-B-j0dWDA.mjs";
import "./scanner-DaG82zNc.mjs";
import "./workspace-patterns-BffPlZ1D.mjs";
import "./project-context-BRIMzhi4.mjs";
import "./resolver-D9cjsz11.mjs";
import { t as fail } from "./cli-error-BrMXlbtx.mjs";
import { i as resolveIntentSkill, r as loadIntentSkill, t as IntentCoreError } from "./core-qlBI_aQL.mjs";
import { a as printDebugInfo, n as coreOptionsFromGlobalFlags } from "./cli-support-BNDR3l1g.mjs";

//#region src/commands/load.ts
function printLoadDebug(loaded) {
	if (!loaded.debug) return;
	printDebugInfo("intent load", [
		["cwd", loaded.debug.cwd],
		["scope", loaded.debug.scope],
		["resolution", loaded.debug.resolution],
		["excludes", loaded.debug.excludes],
		["package", loaded.debug.packageName],
		["version", loaded.debug.version],
		["source", loaded.debug.source],
		["skill", loaded.debug.skillName],
		["path", loaded.debug.path],
		["warnings", loaded.debug.warningCount],
		["packageJsonReadCount", loaded.debug.scan.packageJsonReadCount],
		["packageJsonCacheHits", loaded.debug.scan.packageJsonCacheHits]
	]);
}
async function runLoadCommand(use, options, _scanIntentsOrFail) {
	if (!use) fail("Missing skill use. Expected: intent load <package>#<skill>");
	if (options.json && options.path) fail("Use either --json or --path, not both.");
	const coreOptions = coreOptionsFromGlobalFlags(options);
	if (options.path) {
		let resolved;
		try {
			resolved = resolveIntentSkill(use, coreOptions);
		} catch (err) {
			if (err instanceof IntentCoreError) fail(err.message);
			throw err;
		}
		printLoadDebug(resolved);
		console.log(resolved.path);
		for (const warning of resolved.warnings) console.error(`Warning: ${warning}`);
		return;
	}
	let loaded;
	try {
		loaded = loadIntentSkill(use, coreOptions);
	} catch (err) {
		if (err instanceof IntentCoreError) fail(err.message);
		throw err;
	}
	printLoadDebug(loaded);
	if (options.json) {
		console.log(JSON.stringify({
			package: loaded.packageName,
			skill: loaded.skillName,
			path: loaded.path,
			packageRoot: loaded.packageRoot,
			source: loaded.source,
			version: loaded.version,
			content: loaded.content,
			warnings: loaded.warnings
		}, null, 2));
		return;
	}
	process.stdout.write(loaded.content);
	for (const warning of loaded.warnings) console.error(`Warning: ${warning}`);
}

//#endregion
export { runLoadCommand };
