// @tanstack/intent@0.0.44/dist/project-context-BRIMzhi4.mjs
import { a as readWorkspacePatterns, r as findWorkspaceRoot } from "./workspace-patterns-BffPlZ1D.mjs";
import { existsSync, statSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";

//#region src/core/project-context.ts
/**
* Resolves project structure by walking up from targetPath (or cwd) to find the
* owning package.json, then searches for a workspace root from the package root.
* Falls back to searching from cwd when targetPath points deep into a package.
*/
function resolveProjectContext({ cwd, targetPath }) {
	const resolvedCwd = resolve(cwd);
	const resolvedTargetPath = targetPath ? resolve(resolvedCwd, targetPath) : resolvedCwd;
	const packageRoot = findOwningPackageRoot(resolvedTargetPath);
	const workspaceRoot = findWorkspaceRoot(packageRoot ?? resolvedTargetPath) ?? findWorkspaceRoot(resolvedCwd);
	const workspacePatterns = workspaceRoot ? readWorkspacePatterns(workspaceRoot) ?? [] : [];
	return {
		cwd: resolvedCwd,
		workspaceRoot,
		packageRoot,
		isMonorepo: workspaceRoot !== null,
		workspacePatterns,
		targetPackageJsonPath: packageRoot ? join(packageRoot, "package.json") : null,
		targetSkillsDir: resolveTargetSkillsDir(resolvedTargetPath, packageRoot)
	};
}
function findOwningPackageRoot(startPath) {
	let dir = toSearchDir(startPath);
	let prev;
	while (dir !== prev) {
		if (existsSync(join(dir, "package.json"))) return dir;
		prev = dir;
		dir = dirname(dir);
	}
	return null;
}
function toSearchDir(path) {
	if (!existsSync(path)) return path;
	return statSync(path).isDirectory() ? path : dirname(path);
}
function resolveTargetSkillsDir(targetPath, packageRoot) {
	if (!packageRoot) return null;
	const packageSkillsDir = join(packageRoot, "skills");
	if (isWithinOrEqual(targetPath, packageSkillsDir)) return packageSkillsDir;
	if (targetPath === packageRoot && existsSync(packageSkillsDir)) return packageSkillsDir;
	return null;
}
function isWithinOrEqual(path, parentDir) {
	const rel = relative(parentDir, path);
	return rel === "" || !rel.startsWith("..") && !rel.startsWith("/");
}

//#endregion
export { resolveProjectContext as t };
