// @tanstack/intent@0.0.44/dist/resolver-D9cjsz11.mjs
import { i as parseSkillUse } from "./skill-use-umYvZl94.mjs";
import { t as resolveProjectContext } from "./project-context-BRIMzhi4.mjs";
import { readFileSync } from "node:fs";
import { dirname, isAbsolute, join, relative, resolve } from "node:path";

//#region src/core/package-json.ts
function readPackageJson(dir) {
	try {
		return JSON.parse(readFileSync(join(dir, "package.json"), "utf8"));
	} catch {
		return null;
	}
}

//#endregion
//#region src/core/excludes.ts
const MAX_EXCLUDE_PATTERN_LENGTH = 200;
const PACKAGE_NAME_BOUNDARY = /[^a-zA-Z0-9_.-]/;
function normalizeExcludePatterns(value) {
	if (!Array.isArray(value)) return [];
	return value.filter((pattern) => typeof pattern === "string").map((pattern) => pattern.trim()).filter(Boolean);
}
function isWithinOrEqual(path, parentDir) {
	const rel = relative(parentDir, path);
	return rel === "" || !rel.startsWith("..") && !isAbsolute(rel);
}
function readPackageExcludes(dir) {
	const intent = readPackageJson(dir)?.intent;
	if (!intent || typeof intent !== "object") return [];
	return normalizeExcludePatterns(intent.exclude);
}
function getConfigExcludePatterns(cwd, context = resolveProjectContext({ cwd })) {
	const root = context.workspaceRoot ?? context.packageRoot ?? cwd;
	const dirs = [];
	let dir = cwd;
	while (isWithinOrEqual(dir, root)) {
		dirs.push(dir);
		if (dir === root) break;
		const next = dirname(dir);
		if (next === dir) break;
		dir = next;
	}
	return dirs.reverse().flatMap(readPackageExcludes);
}
function getEffectiveExcludePatterns(options = {}, context) {
	return [...getConfigExcludePatterns(context?.cwd ?? resolve(process.cwd(), options.cwd ?? process.cwd()), context), ...normalizeExcludePatterns(options.exclude)];
}
function normalizeGlobPattern(pattern) {
	if (pattern.length > MAX_EXCLUDE_PATTERN_LENGTH) throw new Error(`Intent exclude pattern is too long: ${pattern.length} characters. Maximum is ${MAX_EXCLUDE_PATTERN_LENGTH}.`);
	return pattern.replace(/\*+/g, "*");
}
function globToRegExp(pattern) {
	const source = normalizeGlobPattern(pattern).split("*").map((part) => part.replace(/[\\^$+?.()|[\]{}]/g, "\\$&")).join(".*");
	return /* @__PURE__ */ new RegExp(`^${source}$`);
}
function compileExcludePatterns(patterns) {
	return patterns.map((pattern) => {
		if (!pattern.includes("*")) {
			normalizeGlobPattern(pattern);
			return {
				pattern,
				matches: (packageName) => packageName === pattern
			};
		}
		const regex = globToRegExp(pattern);
		return {
			pattern,
			matches: (packageName) => regex.test(packageName)
		};
	});
}
function isPackageExcluded(packageName, matchers) {
	return matchers.some((matcher) => matcher.matches(packageName));
}
function warningMentionsPackage(warning, packageName) {
	let idx = warning.indexOf(packageName);
	while (idx !== -1) {
		const before = warning[idx - 1];
		const after = warning[idx + packageName.length];
		if ((before === void 0 || PACKAGE_NAME_BOUNDARY.test(before)) && (after === void 0 || PACKAGE_NAME_BOUNDARY.test(after))) return true;
		idx = warning.indexOf(packageName, idx + packageName.length);
	}
	return false;
}

//#endregion
//#region src/resolver.ts
var ResolveSkillUseError = class extends Error {
	constructor({ availablePackages = [], availableSkills = [], code, packageName, skillName, suggestedSkills = [], use }) {
		super(formatResolveSkillUseErrorMessage({
			availablePackages,
			availableSkills,
			code,
			packageName,
			skillName,
			suggestedSkills,
			use
		}));
		this.name = "ResolveSkillUseError";
		this.availablePackages = availablePackages;
		this.availableSkills = availableSkills;
		this.suggestedSkills = suggestedSkills;
		this.code = code;
		this.packageName = packageName;
		this.skillName = skillName;
		this.use = use;
	}
};
function isResolveSkillUseError(error) {
	return error instanceof ResolveSkillUseError;
}
function getPackageShortName(packageName) {
	return packageName.split("/").pop() ?? packageName;
}
function getPackagePrefixedSkillAlias(packageName, skillName) {
	const prefix = `${getPackageShortName(packageName)}/`;
	return skillName.startsWith(prefix) ? skillName.slice(prefix.length) : null;
}
function getSuggestedSkills(packageName, skillName, skills) {
	const lowerSkillName = skillName.toLowerCase();
	const suggestions = [];
	const seen = /* @__PURE__ */ new Set();
	for (const skill of skills) {
		const alias = getPackagePrefixedSkillAlias(packageName, skill.name);
		const lowerName = skill.name.toLowerCase();
		const lowerAlias = alias?.toLowerCase();
		if (!(lowerAlias === lowerSkillName || lowerName.includes(lowerSkillName) || lowerAlias?.includes(lowerSkillName)) || seen.has(skill.name)) continue;
		seen.add(skill.name);
		suggestions.push(skill.name);
	}
	return suggestions.slice(0, 3);
}
function resolveSkillEntry(packageName, skillName, skills) {
	const exact = skills.find((candidate) => candidate.name === skillName);
	if (exact) return {
		skill: exact,
		suggestedSkills: []
	};
	const aliasMatches = skills.filter((candidate) => getPackagePrefixedSkillAlias(packageName, candidate.name) === skillName);
	if (aliasMatches.length === 1) return {
		skill: aliasMatches[0],
		suggestedSkills: []
	};
	return {
		skill: null,
		suggestedSkills: getSuggestedSkills(packageName, skillName, skills)
	};
}
function resolveSkillUse(use, scanResult) {
	const { packageName, skillName } = parseSkillUse(use);
	const packages = scanResult.packages.filter((pkg$1) => pkg$1.name === packageName);
	const pkg = packages.find((candidate) => candidate.source === "local") ?? packages[0];
	if (!pkg) throw new ResolveSkillUseError({
		availablePackages: scanResult.packages.map((candidate) => candidate.name),
		code: "package-not-found",
		packageName,
		skillName,
		use
	});
	const resolvedSkill = resolveSkillEntry(packageName, skillName, pkg.skills);
	const skill = resolvedSkill.skill;
	if (!skill) throw new ResolveSkillUseError({
		availableSkills: pkg.skills.map((candidate) => candidate.name),
		code: "skill-not-found",
		packageName,
		skillName,
		suggestedSkills: resolvedSkill.suggestedSkills,
		use
	});
	const conflict = scanResult.conflicts.find((candidate) => candidate.packageName === packageName) ?? null;
	return {
		packageName,
		skillName: skill.name,
		path: skill.path,
		source: pkg.source,
		version: pkg.version,
		packageRoot: pkg.packageRoot,
		warnings: scanResult.warnings.filter((warning) => warningMentionsPackage(warning, packageName)),
		conflict
	};
}
function formatResolveSkillUseErrorMessage({ availablePackages, availableSkills, code, packageName, skillName, suggestedSkills, use }) {
	switch (code) {
		case "package-not-found": return `Cannot resolve skill use "${use}": package "${packageName}" was not found.${availablePackages.length > 0 ? ` Available packages: ${availablePackages.join(", ")}.` : ""}`;
		case "skill-not-found": return `Cannot resolve skill use "${use}": skill "${skillName}" was not found in package "${packageName}".${suggestedSkills.length > 0 ? ` Did you mean ${formatSkillSuggestions(packageName, suggestedSkills)}?` : ""}${availableSkills.length > 0 ? ` Available skills: ${availableSkills.join(", ")}.` : ""}`;
	}
}
function formatSkillSuggestions(packageName, skillNames) {
	const uses = skillNames.map((skillName) => `${packageName}#${skillName}`);
	if (uses.length <= 2) return uses.join(" or ");
	return `${uses.slice(0, -1).join(", ")}, or ${uses.at(-1)}`;
}

//#endregion
export { compileExcludePatterns as a, warningMentionsPackage as c, resolveSkillUse as i, isResolveSkillUseError as n, getEffectiveExcludePatterns as o, resolveSkillEntry as r, isPackageExcluded as s, ResolveSkillUseError as t };
