// @tanstack/intent@0.0.44/dist/scanner-DaG82zNc.mjs
import { a as listNestedNodeModulesPackageDirs, c as parseFrontmatter, i as getDeps, l as resolveDepDir, n as detectGlobalNodeModules, o as listNodeModulesPackageDirs, r as findSkillFiles, s as nodeReadFs, t as createFsIdentityCache, u as toPosixPath } from "./utils-9fhWAVua.mjs";
import { r as rewriteSkillLoadPaths } from "./skill-paths-B-j0dWDA.mjs";
import { n as findWorkspacePackages, r as findWorkspaceRoot } from "./workspace-patterns-BffPlZ1D.mjs";
import { t as detectPackageManager } from "./package-manager-BUgTjW9Q.mjs";
import { createRequire } from "node:module";
import { existsSync } from "node:fs";
import { dirname, isAbsolute, join, relative, resolve, sep } from "node:path";
import semver from "semver";

//#region src/discovery/register.ts
function isLocalToProject(dirPath, projectRoot) {
	return dirPath.startsWith(projectRoot + sep) || dirPath.startsWith(projectRoot + "/");
}
function createPackageRegistrar(opts) {
	const attemptedPackageRoots = /* @__PURE__ */ new Set();
	const scannedNodeModulesDirs = /* @__PURE__ */ new Set();
	function shouldAttemptPackageRoot(dirPath) {
		const key = opts.getFsIdentity(dirPath);
		if (attemptedPackageRoots.has(key)) return false;
		attemptedPackageRoots.add(key);
		return true;
	}
	function scanNodeModulesDir(nodeModulesDir, source = "local") {
		if (!existsSync(nodeModulesDir)) return;
		const key = opts.getFsIdentity(nodeModulesDir);
		if (scannedNodeModulesDirs.has(key)) return;
		scannedNodeModulesDirs.add(key);
		for (const dirPath of listNodeModulesPackageDirs(nodeModulesDir)) tryRegister(dirPath, "unknown", source);
	}
	function scanTarget(target, source = "local") {
		if (!target.path || !target.exists || target.scanned) return;
		target.scanned = true;
		scanNodeModulesDir(target.path, source);
	}
	function tryRegister(dirPath, fallbackName, source = "local") {
		if (!shouldAttemptPackageRoot(dirPath)) return false;
		const skillsDir = join(dirPath, "skills");
		if (!opts.exists(skillsDir)) return false;
		const pkgJson = opts.readPkgJson(dirPath);
		if (!pkgJson) {
			opts.warnings.push(`Could not read package.json for ${dirPath}`);
			return false;
		}
		const name = typeof pkgJson.name === "string" ? pkgJson.name : fallbackName;
		const version = typeof pkgJson.version === "string" ? pkgJson.version : "0.0.0";
		const intent = opts.validateIntentField(name, pkgJson.intent) ?? opts.deriveIntentConfig(pkgJson);
		if (!intent) {
			opts.warnings.push(`${name} has a skills/ directory but could not determine repo/docs from package.json (add a "repository" field or explicit "intent" config)`);
			return false;
		}
		const skills = opts.discoverSkills(skillsDir, name);
		if (isLocalToProject(dirPath, opts.projectRoot)) rewriteSkillLoadPaths({
			packageName: name,
			packageRoot: dirPath,
			projectRoot: opts.projectRoot,
			skills
		});
		const candidate = {
			name,
			version,
			intent,
			skills,
			packageRoot: dirPath,
			source
		};
		const existingIndex = opts.packageIndexes.get(name);
		if (existingIndex === void 0) {
			opts.rememberVariant(candidate);
			opts.packageIndexes.set(name, opts.packages.push(candidate) - 1);
			return true;
		}
		const existing = opts.packages[existingIndex];
		if (existing.packageRoot === candidate.packageRoot) return false;
		opts.rememberVariant(existing);
		opts.rememberVariant(candidate);
		const existingDepth = opts.getPackageDepth(existing.packageRoot, opts.projectRoot);
		const candidateDepth = opts.getPackageDepth(candidate.packageRoot, opts.projectRoot);
		if (candidateDepth < existingDepth || candidateDepth === existingDepth && opts.comparePackageVersions(candidate.version, existing.version) > 0) opts.packages[existingIndex] = candidate;
		return true;
	}
	return {
		scanNodeModulesDir,
		scanTarget,
		tryRegister
	};
}

//#endregion
//#region src/discovery/walk.ts
function createDependencyWalker(opts) {
	const walkVisited = /* @__PURE__ */ new Set();
	const depDirCache = /* @__PURE__ */ new Map();
	function resolveDepDirCached(depName, fromDir) {
		const fromKey = opts.getFsIdentity(fromDir);
		let byDepName = depDirCache.get(fromKey);
		if (!byDepName) {
			byDepName = /* @__PURE__ */ new Map();
			depDirCache.set(fromKey, byDepName);
		}
		if (!byDepName.has(depName)) byDepName.set(depName, resolveDepDir(depName, fromDir));
		return byDepName.get(depName) ?? null;
	}
	function walkDepsOf(pkgJson, fromDir, includeDevDeps = false) {
		for (const depName of getDeps(pkgJson, includeDevDeps)) {
			const depDir = resolveDepDirCached(depName, fromDir);
			if (!depDir) continue;
			opts.tryRegister(depDir, depName);
			walkDeps(depDir, depName);
		}
	}
	function walkDeps(pkgDir, pkgName) {
		const pkgKey = opts.getFsIdentity(pkgDir);
		if (walkVisited.has(pkgKey)) return;
		walkVisited.add(pkgKey);
		const pkgJson = opts.readPkgJson(pkgDir);
		if (!pkgJson) {
			opts.warnings.push(`Could not read package.json for ${pkgName} (skipping dependency walk)`);
			return;
		}
		walkDepsOf(pkgJson, pkgKey);
	}
	function walkKnownPackages() {
		for (const pkg of [...opts.packages]) walkDeps(pkg.packageRoot, pkg.name);
	}
	function walkProjectDeps() {
		const projectPkg = readPkgJsonWithWarning(opts.projectRoot, "project");
		if (!projectPkg) return;
		walkDepsOf(projectPkg, opts.projectRoot, true);
	}
	function readPkgJsonWithWarning(dirPath, label) {
		const result = opts.fsCache.readPackageJsonResult(dirPath);
		if (!result.packageJson) {
			if (result.error?.code !== "ENOENT") opts.warnings.push(`Could not read ${label} package.json at ${dirPath}: ${result.error instanceof Error ? result.error.message : "invalid package.json"}`);
			return null;
		}
		return result.packageJson;
	}
	function walkWorkspacePackages() {
		for (const wsDir of findWorkspacePackages(opts.projectRoot)) {
			opts.scanNodeModulesDir(join(wsDir, "node_modules"));
			const wsPkg = readPkgJsonWithWarning(wsDir, "workspace");
			if (wsPkg) walkDepsOf(wsPkg, wsDir);
		}
	}
	function scanNestedNodeModulesDir(nodeModulesDir) {
		for (const dirPath of listNestedNodeModulesPackageDirs(nodeModulesDir, opts.getFsIdentity)) {
			if (!opts.tryRegister(dirPath, "unknown")) continue;
			const pkgJson = opts.readPkgJson(dirPath);
			walkDeps(dirPath, typeof pkgJson?.name === "string" ? pkgJson.name : "unknown");
		}
	}
	return {
		scanNestedNodeModulesDir,
		walkKnownPackages,
		walkProjectDeps,
		walkWorkspacePackages
	};
}

//#endregion
//#region src/fs-cache.ts
function isRecord(value) {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}
function createIntentFsCache() {
	const packageJsonCache = /* @__PURE__ */ new Map();
	const skillFilesCache = /* @__PURE__ */ new Map();
	let activeFs = nodeReadFs;
	const getFsIdentity = createFsIdentityCache(() => activeFs);
	const stats = {
		packageJsonReadCount: 0,
		packageJsonCacheHits: 0
	};
	function readPackageJsonResult(dir) {
		const key = getFsIdentity(dir);
		const cached = packageJsonCache.get(key);
		if (cached) {
			stats.packageJsonCacheHits += 1;
			return cached;
		}
		stats.packageJsonReadCount += 1;
		try {
			const parsed = JSON.parse(activeFs.readFileSync(join(dir, "package.json"), "utf8"));
			const result = {
				packageJson: isRecord(parsed) ? parsed : null,
				error: null
			};
			packageJsonCache.set(key, result);
			return result;
		} catch (error) {
			const result = {
				packageJson: null,
				error
			};
			packageJsonCache.set(key, result);
			return result;
		}
	}
	function readPackageJson(dir) {
		return readPackageJsonResult(dir).packageJson;
	}
	function findSkillFiles$1(dir) {
		const key = getFsIdentity(dir);
		const cached = skillFilesCache.get(key);
		if (cached) return [...cached];
		const files = findSkillFiles(dir, activeFs);
		skillFilesCache.set(key, files);
		return [...files];
	}
	return {
		readPackageJson,
		readPackageJsonResult,
		findSkillFiles: findSkillFiles$1,
		getFsIdentity,
		getStats: () => ({ ...stats }),
		useFs: (fs) => {
			activeFs = fs;
		},
		getReadFs: () => activeFs,
		exists: (path) => activeFs.existsSync(path)
	};
}

//#endregion
//#region src/scanner.ts
const requireFromHere = createRequire(import.meta.url);
function findPnpFile(start) {
	let dir = resolve(start);
	let prev;
	while (dir !== prev) {
		for (const fileName of [".pnp.cjs", ".pnp.js"]) {
			const pnpPath = join(dir, fileName);
			if (existsSync(pnpPath)) return pnpPath;
		}
		prev = dir;
		dir = dirname(dir);
	}
	return null;
}
function assertLocalNodeModulesSupported(root) {
	if (existsSync(join(root, "deno.json")) && !existsSync(join(root, "node_modules"))) throw new Error("Deno without node_modules is not yet supported. Add `\"nodeModulesDir\": \"auto\"` to your deno.json to use intent.");
}
function loadPnpApi(root) {
	const pnpPath = findPnpFile(root);
	if (!pnpPath) return null;
	const moduleApi = requireFromHere("node:module");
	const originalResolveFilename = moduleApi._resolveFilename;
	const readFs = requireFromHere("node:fs");
	try {
		const pnpModule = requireFromHere(pnpPath);
		if (typeof pnpModule.setup === "function") pnpModule.setup();
		moduleApi._resolveFilename = originalResolveFilename;
		if (typeof pnpModule.getDependencyTreeRoots === "function" && typeof pnpModule.getPackageInformation === "function") return {
			api: pnpModule,
			readFs
		};
		return {
			api: createRequire(join(dirname(pnpPath), "package.json"))("pnpapi"),
			readFs
		};
	} catch (err) {
		moduleApi._resolveFilename = originalResolveFilename;
		try {
			const foundApi = moduleApi.findPnpApi?.(root);
			if (foundApi) return {
				api: foundApi,
				readFs
			};
		} catch {}
		const msg = err instanceof Error ? err.message : String(err);
		throw new Error(`Yarn PnP project detected, but Intent could not load Yarn's PnP API from ${pnpPath}: ${msg}`);
	}
}
function getPnpLocatorKey(locator) {
	return `${locator.name ?? "<top>"}@${locator.reference ?? "<top>"}`;
}
function getPnpDependencyLocator(dependencyName, target) {
	if (target === null) return null;
	if (Array.isArray(target)) return {
		name: target[0],
		reference: target[1]
	};
	return {
		name: dependencyName,
		reference: target
	};
}
function validateIntentField(_pkgName, intent) {
	if (!intent || typeof intent !== "object") return null;
	const pb = intent;
	if (pb.version !== 1) return null;
	if (typeof pb.repo !== "string" || !pb.repo) return null;
	if (typeof pb.docs !== "string" || !pb.docs) return null;
	const requires = Array.isArray(pb.requires) ? pb.requires.filter((r) => typeof r === "string") : void 0;
	return {
		version: 1,
		repo: pb.repo,
		docs: pb.docs,
		requires
	};
}
/**
* Derive an IntentConfig from standard package.json fields when no explicit
* `intent` field is present. A package with a `skills/` directory signals
* intent support; `repo` and `docs` are derived from `repository` and
* `homepage`.
*/
function deriveIntentConfig(pkgJson) {
	let repo = null;
	if (typeof pkgJson.repository === "string") repo = pkgJson.repository;
	else if (pkgJson.repository && typeof pkgJson.repository === "object" && typeof pkgJson.repository.url === "string") {
		repo = pkgJson.repository.url;
		repo = repo.replace(/^git\+/, "").replace(/\.git$/, "").replace(/^https?:\/\/github\.com\//, "");
	}
	const docs = typeof pkgJson.homepage === "string" ? pkgJson.homepage : void 0;
	if (!repo) return null;
	const intentPartial = pkgJson.intent;
	const requires = intentPartial && Array.isArray(intentPartial.requires) ? intentPartial.requires.filter((r) => typeof r === "string") : void 0;
	return {
		version: 1,
		repo,
		docs: docs ?? "",
		requires
	};
}
function readSkillEntry(skillsDir, childDir, skillFile, readFs = nodeReadFs) {
	const fm = parseFrontmatter(skillFile, readFs);
	const relName = toPosixPath(relative(skillsDir, childDir));
	const desc = typeof fm?.description === "string" ? fm.description.replace(/\s+/g, " ").trim() : "";
	return {
		name: typeof fm?.name === "string" ? fm.name : relName,
		path: skillFile,
		description: desc,
		type: typeof fm?.type === "string" ? fm.type : void 0,
		framework: typeof fm?.framework === "string" ? fm.framework : void 0
	};
}
function discoverSkillByNameHint(skillsDir, packageName, skillNameHint, readFs = nodeReadFs) {
	const skills = [];
	const seen = /* @__PURE__ */ new Set();
	const skillNameHints = getSkillNameHints(packageName, skillNameHint);
	for (const hint of skillNameHints) {
		const resolvedHint = resolveSkillNameHintPath(skillsDir, hint);
		if (!resolvedHint) continue;
		const { childDir, skillFile } = resolvedHint;
		if (!readFs.existsSync(skillFile)) continue;
		const skill = readSkillEntry(skillsDir, childDir, skillFile, readFs);
		if (skill.name !== hint || seen.has(skill.name)) continue;
		seen.add(skill.name);
		skills.push(skill);
	}
	return skills;
}
function discoverSkills(skillsDir, fsCache) {
	const readFs = fsCache.getReadFs();
	return fsCache.findSkillFiles(skillsDir).flatMap((skillFile) => {
		const childDir = dirname(skillFile);
		if (childDir === skillsDir) return [];
		return [readSkillEntry(skillsDir, childDir, skillFile, readFs)];
	});
}
function getPackageShortName(packageName) {
	return packageName.split("/").pop() ?? packageName;
}
function isWithinOrEqual(path, parentDir) {
	const rel = relative(parentDir, path);
	return rel === "" || !rel.startsWith("..") && !isAbsolute(rel);
}
function resolveSkillNameHintPath(skillsDir, hint) {
	if (hint.startsWith("/") || hint.startsWith("\\")) return null;
	const parts = hint.split("/");
	if (parts.some((part) => part === "" || part === "." || part === ".." || part.includes("\\"))) return null;
	const resolvedSkillsDir = resolve(skillsDir);
	const childDir = resolve(resolvedSkillsDir, ...parts);
	if (!isWithinOrEqual(childDir, resolvedSkillsDir)) return null;
	return {
		childDir,
		skillFile: join(childDir, "SKILL.md")
	};
}
function getSkillNameHints(packageName, skillNameHint) {
	const packageShortName = getPackageShortName(packageName);
	if (skillNameHint.startsWith(`${packageShortName}/`)) return [skillNameHint];
	return [skillNameHint, `${packageShortName}/${skillNameHint}`];
}
function topoSort(packages) {
	const byName = new Map(packages.map((p) => [p.name, p]));
	const visited = /* @__PURE__ */ new Set();
	const sorted = [];
	function visit(name) {
		if (visited.has(name)) return;
		visited.add(name);
		const pkg = byName.get(name);
		if (!pkg) return;
		for (const dep of pkg.intent.requires ?? []) visit(dep);
		sorted.push(pkg);
	}
	for (const pkg of packages) visit(pkg.name);
	return sorted;
}
function getPackageDepth(packageRoot, projectRoot) {
	return relative(projectRoot, packageRoot).split(sep).length;
}
function normalizeVersion(version) {
	const validVersion = semver.valid(version);
	if (validVersion) return validVersion;
	return semver.coerce(version)?.version ?? null;
}
function comparePackageVersions(a, b) {
	const versionA = normalizeVersion(a);
	const versionB = normalizeVersion(b);
	if (!versionA || !versionB) {
		if (versionA) return 1;
		if (versionB) return -1;
		return 0;
	}
	return semver.compare(versionA, versionB);
}
function formatVariantWarning(name, variants, chosen) {
	const uniqueVersions = new Set(variants.map((variant) => variant.version));
	if (uniqueVersions.size <= 1) return null;
	const details = variants.map((variant) => `${variant.version} at ${variant.packageRoot}`).join(", ");
	return `Found ${variants.length} installed variants of ${name} across ${uniqueVersions.size} versions (${details}). Using ${chosen.version} from ${chosen.packageRoot}.`;
}
function toVersionConflict(packageName, variants, chosen) {
	if (new Set(variants.map((variant) => variant.version)).size <= 1) return null;
	return {
		packageName,
		chosen: {
			version: chosen.version,
			packageRoot: chosen.packageRoot
		},
		variants
	};
}
function getScanScope(options) {
	return options.scope ?? (options.includeGlobal ? "local-and-global" : "local");
}
function scanForIntents(root, options = {}) {
	const projectRoot = root ?? process.cwd();
	const scanScope = getScanScope(options);
	const fsCache = options.fsCache ?? createIntentFsCache();
	const workspaceRoot = findWorkspaceRoot(projectRoot);
	const packageManager = detectPackageManager(projectRoot, [workspaceRoot]);
	const nodeModulesDir = join(projectRoot, "node_modules");
	const explicitGlobalNodeModules = process.env.INTENT_GLOBAL_NODE_MODULES?.trim() || null;
	const packages = [];
	const warnings = [];
	const conflicts = [];
	const nodeModules = {
		local: {
			path: nodeModulesDir,
			detected: true,
			exists: existsSync(nodeModulesDir),
			scanned: false
		},
		global: {
			path: explicitGlobalNodeModules,
			detected: Boolean(explicitGlobalNodeModules),
			exists: explicitGlobalNodeModules ? existsSync(explicitGlobalNodeModules) : false,
			scanned: false,
			source: explicitGlobalNodeModules ? "INTENT_GLOBAL_NODE_MODULES" : void 0
		}
	};
	const packageIndexes = /* @__PURE__ */ new Map();
	const packageVariants = /* @__PURE__ */ new Map();
	let pnpApi;
	function getPnpApi() {
		if (scanScope === "global") return null;
		if (pnpApi === void 0) {
			const loaded = loadPnpApi(projectRoot);
			pnpApi = loaded?.api ?? null;
			if (loaded) fsCache.useFs(loaded.readFs);
		}
		return pnpApi;
	}
	function getStats() {
		return fsCache.getStats();
	}
	function rememberVariant(pkg) {
		let variants = packageVariants.get(pkg.name);
		if (!variants) {
			variants = /* @__PURE__ */ new Map();
			packageVariants.set(pkg.name, variants);
		}
		variants.set(pkg.packageRoot, {
			version: pkg.version,
			packageRoot: pkg.packageRoot
		});
	}
	function ensureGlobalNodeModules() {
		if (!nodeModules.global.path && !explicitGlobalNodeModules) {
			const detected = detectGlobalNodeModules(packageManager);
			nodeModules.global.path = detected.path;
			nodeModules.global.source = detected.source;
			nodeModules.global.detected = Boolean(detected.path);
			nodeModules.global.exists = detected.path ? existsSync(detected.path) : false;
		}
	}
	function readPkgJson(dirPath) {
		return fsCache.readPackageJson(dirPath);
	}
	const { scanNodeModulesDir, scanTarget, tryRegister } = createPackageRegistrar({
		comparePackageVersions,
		deriveIntentConfig,
		discoverSkills: (skillsDir) => discoverSkills(skillsDir, fsCache),
		getPackageDepth,
		getFsIdentity: fsCache.getFsIdentity,
		exists: fsCache.exists,
		packageIndexes,
		packages,
		projectRoot,
		readPkgJson,
		rememberVariant,
		validateIntentField,
		warnings
	});
	const { scanNestedNodeModulesDir, walkKnownPackages, walkProjectDeps, walkWorkspacePackages } = createDependencyWalker({
		fsCache,
		getFsIdentity: fsCache.getFsIdentity,
		packages,
		projectRoot,
		readPkgJson,
		scanNodeModulesDir,
		tryRegister,
		warnings
	});
	function scanPnpPackages(api) {
		const visited = /* @__PURE__ */ new Set();
		const projectLocator = api.findPackageLocator?.(projectRoot.endsWith(sep) ? projectRoot : `${projectRoot}${sep}`);
		const roots = workspaceRoot && workspaceRoot !== projectRoot && projectLocator ? [projectLocator] : api.getDependencyTreeRoots?.() ?? (api.topLevel ? [api.topLevel] : []);
		function visit(locator) {
			const key = getPnpLocatorKey(locator);
			if (visited.has(key)) return;
			visited.add(key);
			const info = api.getPackageInformation(locator);
			if (!info) return;
			tryRegister(info.packageLocation.replace(/[\\/]$/, ""), locator.name ?? "unknown");
			for (const [dependencyName, target] of info.packageDependencies) {
				const dependencyLocator = getPnpDependencyLocator(dependencyName, target);
				if (dependencyLocator) visit(dependencyLocator);
			}
		}
		for (const locator of roots) visit(locator);
	}
	function scanLocalPackages() {
		if (!nodeModules.local.exists) {
			const api = getPnpApi();
			if (api) {
				scanPnpPackages(api);
				return;
			}
		}
		assertLocalNodeModulesSupported(projectRoot);
		const packageCountBeforeLocalDiscovery = packages.length;
		walkWorkspacePackages();
		const packageCountBeforeDependencyDiscovery = packages.length;
		scanTarget(nodeModules.local);
		walkKnownPackages();
		walkProjectDeps();
		const shouldTryPnpFallback = packages.length === packageCountBeforeDependencyDiscovery;
		if (nodeModules.local.path && nodeModules.local.exists && packages.length === packageCountBeforeLocalDiscovery) scanNestedNodeModulesDir(nodeModules.local.path);
		if (shouldTryPnpFallback) {
			const api = getPnpApi();
			if (api) scanPnpPackages(api);
		}
	}
	function scanGlobalPackages() {
		ensureGlobalNodeModules();
		scanTarget(nodeModules.global, "global");
	}
	switch (scanScope) {
		case "local":
			scanLocalPackages();
			break;
		case "local-and-global":
			scanLocalPackages();
			scanGlobalPackages();
			walkKnownPackages();
			walkProjectDeps();
			break;
		case "global":
			scanGlobalPackages();
			break;
	}
	if (!nodeModules.local.exists && !nodeModules.global.exists) return {
		packageManager,
		packages,
		warnings,
		conflicts,
		nodeModules,
		stats: getStats()
	};
	for (const pkg of packages) {
		const variants = packageVariants.get(pkg.name);
		if (!variants) continue;
		const conflict = toVersionConflict(pkg.name, [...variants.values()], pkg);
		if (conflict) conflicts.push(conflict);
		const warning = formatVariantWarning(pkg.name, [...variants.values()], pkg);
		if (warning) warnings.push(warning);
	}
	return {
		packageManager,
		packages: topoSort(packages),
		warnings,
		conflicts,
		nodeModules,
		stats: getStats()
	};
}
function scanIntentPackageAtRoot(packageRoot, options = {}) {
	const projectRoot = options.projectRoot ?? packageRoot;
	const packages = [];
	const warnings = [];
	const packageIndexes = /* @__PURE__ */ new Map();
	const fsCache = options.fsCache ?? createIntentFsCache();
	function readPkgJson(dirPath) {
		return fsCache.readPackageJson(dirPath);
	}
	const { tryRegister } = createPackageRegistrar({
		comparePackageVersions,
		deriveIntentConfig,
		discoverSkills: options.skillNameHint ? (skillsDir, packageName) => discoverSkillByNameHint(skillsDir, packageName, options.skillNameHint, fsCache.getReadFs()) : (skillsDir) => discoverSkills(skillsDir, fsCache),
		getPackageDepth,
		getFsIdentity: fsCache.getFsIdentity,
		exists: fsCache.exists,
		packageIndexes,
		packages,
		projectRoot,
		readPkgJson,
		rememberVariant() {},
		validateIntentField,
		warnings
	});
	tryRegister(packageRoot, options.fallbackName ?? "unknown", options.source ?? "local");
	return {
		package: packages[0] ?? null,
		warnings
	};
}

//#endregion
export { scanIntentPackageAtRoot as n, createIntentFsCache as r, scanForIntents as t };
