// @tanstack/intent@0.0.44/dist/setup-BklJkH_6.mjs
import { a as readWorkspacePatterns, r as findWorkspaceRoot, t as findPackagesWithSkills } from "./workspace-patterns-BffPlZ1D.mjs";
import { t as resolveProjectContext } from "./project-context-BRIMzhi4.mjs";
import { existsSync, mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { basename, join, relative } from "node:path";

//#region src/setup.ts
function isGenericWorkspaceName(name, root) {
	const normalized = name.trim().toLowerCase();
	return normalized.length === 0 || normalized === "unknown" || normalized === "root" || normalized === "workspace" || normalized === "monorepo" || normalized === basename(root).toLowerCase();
}
function deriveWorkspacePackageName(root, repo, packageDirs) {
	const repoName = repo.split("/").filter(Boolean).pop() || basename(root);
	for (const packageDir of packageDirs) {
		const pkgJson = readPackageJson(packageDir);
		const pkgName = typeof pkgJson.name === "string" ? pkgJson.name : null;
		if (pkgName?.startsWith("@")) return `${pkgName.split("/")[0]}/${repoName}`;
	}
	return repoName;
}
function readPackageJson(root) {
	const pkgPath = join(root, "package.json");
	try {
		return JSON.parse(readFileSync(pkgPath, "utf8"));
	} catch (err) {
		if (!(err && typeof err === "object" && "code" in err && err.code === "ENOENT")) console.error(`Warning: could not read ${pkgPath}: ${err instanceof Error ? err.message : err}`);
		return {};
	}
}
function detectRepo(pkgJson, fallback) {
	const intent = pkgJson.intent;
	if (typeof intent?.repo === "string") return intent.repo;
	if (typeof pkgJson.repository === "string") return pkgJson.repository.replace(/^git\+/, "").replace(/\.git$/, "").replace(/^https?:\/\/github\.com\//, "");
	if (pkgJson.repository && typeof pkgJson.repository === "object" && typeof pkgJson.repository.url === "string") return pkgJson.repository.url.replace(/^git\+/, "").replace(/\.git$/, "").replace(/^https?:\/\/github\.com\//, "");
	return fallback;
}
function normalizePattern(pattern) {
	return pattern.endsWith("**") ? pattern : pattern.replace(/\/$/, "") + "/**";
}
function buildWatchPaths(root, packageDirs) {
	const paths = /* @__PURE__ */ new Set();
	if (existsSync(join(root, "docs"))) paths.add("docs/**");
	for (const packageDir of packageDirs) {
		const relDir = relative(root, packageDir).split("\\").join("/");
		if (existsSync(join(packageDir, "src"))) paths.add(`${relDir}/src/**`);
		const intent = readPackageJson(packageDir).intent;
		const docs = typeof intent?.docs === "string" ? intent.docs : "docs/";
		if (!docs.startsWith("http://") && !docs.startsWith("https://")) paths.add(normalizePattern(join(relDir, docs).split("\\").join("/")));
	}
	if (paths.size === 0) {
		paths.add("packages/*/src/**");
		paths.add("packages/*/docs/**");
	}
	return [...paths].sort().map((path) => `      - '${path}'`).join("\n");
}
function detectVars(root, packageDirs) {
	const pkgJson = readPackageJson(root);
	const rawName = typeof pkgJson.name === "string" ? pkgJson.name : "unknown";
	const docs = typeof pkgJson.intent?.docs === "string" ? pkgJson.intent.docs : "docs/";
	const isMonorepo = packageDirs !== void 0;
	const repo = detectRepo(pkgJson, detectRepo((packageDirs?.[0] ? readPackageJson(packageDirs[0]) : null) ?? {}, basename(root)));
	let packageName = rawName;
	if (isMonorepo && isGenericWorkspaceName(rawName, root)) packageName = deriveWorkspacePackageName(root, repo, packageDirs);
	const shortName = packageName.replace(/^@[^/]+\//, "");
	let srcPath = isMonorepo ? "packages/*/src/**" : `packages/${shortName}/src/**`;
	if (!isMonorepo && existsSync(join(root, "src"))) srcPath = "src/**";
	const docsPath = isMonorepo ? "packages/*/docs/**" : docs;
	return {
		PACKAGE_NAME: packageName,
		PACKAGE_LABEL: packageName,
		PAYLOAD_PACKAGE: packageName,
		REPO: repo,
		DOCS_PATH: docsPath.endsWith("**") ? docsPath : docsPath.replace(/\/$/, "") + "/**",
		SRC_PATH: srcPath,
		WATCH_PATHS: isMonorepo ? buildWatchPaths(root, packageDirs) : `      - '${docs.endsWith("**") ? docs : docs.replace(/\/$/, "") + "/**"}'\n      - '${srcPath}'`
	};
}
function applyVars(content, vars) {
	return content.replace(/\{\{PACKAGE_NAME\}\}/g, vars.PACKAGE_NAME).replace(/\{\{PACKAGE_LABEL\}\}/g, vars.PACKAGE_LABEL).replace(/\{\{PAYLOAD_PACKAGE\}\}/g, vars.PAYLOAD_PACKAGE).replace(/\{\{REPO\}\}/g, vars.REPO).replace(/\{\{DOCS_PATH\}\}/g, vars.DOCS_PATH).replace(/\{\{SRC_PATH\}\}/g, vars.SRC_PATH).replace(/\{\{WATCH_PATHS\}\}/g, vars.WATCH_PATHS);
}
function copyTemplates(srcDir, destDir, vars) {
	const copied = [];
	const skipped = [];
	if (!existsSync(srcDir)) return {
		copied,
		skipped
	};
	mkdirSync(destDir, { recursive: true });
	for (const entry of readdirSync(srcDir)) {
		const srcPath = join(srcDir, entry);
		const destPath = join(destDir, entry);
		if (existsSync(destPath)) {
			skipped.push(destPath);
			continue;
		}
		let content = readFileSync(srcPath, "utf8");
		if (vars.WATCH_PATHS.includes("\n")) content = content.replace(/\s+- '?\{\{DOCS_PATH\}\}'?\n\s+- '?\{\{SRC_PATH\}\}'?/, vars.WATCH_PATHS);
		writeFileSync(destPath, applyVars(content, vars));
		copied.push(destPath);
	}
	return {
		copied,
		skipped
	};
}
function runEditPackageJson(root) {
	const result = {
		added: [],
		alreadyPresent: []
	};
	const context = resolveProjectContext({ cwd: root });
	const packageRoot = context.packageRoot ?? root;
	const pkgPath = join(packageRoot, "package.json");
	if (!existsSync(pkgPath)) {
		console.error("No package.json found in " + packageRoot);
		process.exitCode = 1;
		return result;
	}
	const raw = readFileSync(pkgPath, "utf8");
	let pkg;
	try {
		pkg = JSON.parse(raw);
	} catch (err) {
		const detail = err instanceof SyntaxError ? err.message : String(err);
		console.error(`Failed to parse ${pkgPath}: ${detail}`);
		process.exitCode = 1;
		return result;
	}
	const indentMatch = raw.match(/^(\s+)"/m);
	const indentSize = indentMatch?.[1] ? indentMatch[1].length : 2;
	if (!Array.isArray(pkg.keywords)) pkg.keywords = [];
	const keywords = pkg.keywords;
	if (keywords.includes("tanstack-intent")) result.alreadyPresent.push("keywords: \"tanstack-intent\"");
	else {
		keywords.push("tanstack-intent");
		result.added.push("keywords: \"tanstack-intent\"");
	}
	if (!Array.isArray(pkg.files)) pkg.files = [];
	const files = pkg.files;
	const requiredFiles = context.isMonorepo ? ["skills"] : ["skills", "!skills/_artifacts"];
	for (const entry of requiredFiles) if (files.includes(entry)) result.alreadyPresent.push(`files: "${entry}"`);
	else {
		files.push(entry);
		result.added.push(`files: "${entry}"`);
	}
	writeFileSync(pkgPath, JSON.stringify(pkg, null, indentSize) + "\n");
	for (const a of result.added) console.log(`✓ Added ${a}`);
	for (const a of result.alreadyPresent) console.log(`  Already present: ${a}`);
	return result;
}
/**
* When run from a monorepo root, finds all workspace packages with SKILL.md
* files and runs the given command on each. Falls back to single-package
* behavior only when no workspace config is detected. If workspace config
* exists but no packages have skills, warns and returns empty.
*/
function runForEachPackage(root, runOne) {
	const isMonorepo = readWorkspacePatterns(root) !== null;
	const pkgsWithSkills = isMonorepo ? findPackagesWithSkills(root) : [];
	if (!isMonorepo) return runOne(root);
	if (pkgsWithSkills.length === 0) {
		console.log("No workspace packages with skills found.");
		return [];
	}
	return pkgsWithSkills.map((pkgDir) => {
		const rel = relative(root, pkgDir) || ".";
		console.log(`\n── ${rel} ──`);
		return {
			package: rel,
			result: runOne(pkgDir)
		};
	});
}
function runEditPackageJsonAll(root) {
	return runForEachPackage(root, runEditPackageJson);
}
function runSetupGithubActions(root, metaDir) {
	const workspaceRoot = findWorkspaceRoot(root) ?? root;
	const packageDirs = findPackagesWithSkills(workspaceRoot);
	const vars = detectVars(workspaceRoot, packageDirs.length > 0 ? packageDirs : void 0);
	const result = {
		workflows: [],
		skipped: []
	};
	const { copied, skipped } = copyTemplates(join(metaDir, "templates", "workflows"), join(workspaceRoot, ".github", "workflows"), vars);
	result.workflows = copied;
	result.skipped = skipped;
	for (const f of result.workflows) console.log(`✓ Copied workflow: ${f}`);
	for (const f of result.skipped) console.log(`  Already exists: ${f}`);
	if (result.workflows.length === 0 && result.skipped.length === 0) console.log("No templates directory found. Is @tanstack/intent installed?");
	else if (result.workflows.length > 0) {
		console.log(`\nTemplate variables applied:`);
		console.log(`  Package:  ${vars.PACKAGE_LABEL}`);
		console.log(`  Repo:     ${vars.REPO}`);
		console.log(`  Mode:     ${packageDirs.length > 0 ? `monorepo (${packageDirs.length} packages with skills)` : "single package"}`);
	}
	return result;
}

//#endregion
export { runEditPackageJsonAll as n, runSetupGithubActions as r, runEditPackageJson as t };
