// @tanstack/intent@0.0.44/dist/skill-paths-B-j0dWDA.mjs
import { u as toPosixPath } from "./utils-9fhWAVua.mjs";
import { existsSync } from "node:fs";
import { join, relative } from "node:path";

//#region src/skill-paths.ts
function isAbsolutePath(path) {
	return path.startsWith("/") || path.startsWith("\\") || /^[A-Za-z]:[\\/]/.test(path);
}
function getPathSegments(path) {
	return path.replace(/\\/g, "/").split("/");
}
function isPackageManagerInternalSegment(segment) {
	return segment === ".pnpm" || segment === ".bun" || segment === ".yarn";
}
function isStableLoadPath(path) {
	const normalized = path.trim();
	const segments = getPathSegments(normalized);
	return normalized !== "" && !isAbsolutePath(normalized) && !segments.includes("..") && !segments.some(isPackageManagerInternalSegment);
}
function rewriteSkillLoadPaths({ packageName, packageRoot, projectRoot, skills }) {
	const hasStableSymlink = packageName !== "" && existsSync(join(projectRoot, "node_modules", packageName));
	for (const skill of skills) if (hasStableSymlink) skill.path = `node_modules/${packageName}/${toPosixPath(relative(packageRoot, skill.path))}`;
	else skill.path = toPosixPath(relative(projectRoot, skill.path));
}
function formatRuntimeSkillLookupComment(target) {
	return `Runtime lookup only: run \`npx @tanstack/intent@latest load ${target.packageName}#${target.skillName} --path\`, and load its reported path for this session. Do not copy the resolved path into this file.`;
}
function formatRuntimeSkillLookupHint(target) {
	return `Lookup: ${formatRuntimeSkillLookupComment(target)}`;
}

//#endregion
export { isStableLoadPath as n, rewriteSkillLoadPaths as r, formatRuntimeSkillLookupHint as t };
