// @tanstack/intent@0.0.44/dist/skill-use-umYvZl94.mjs
//#region src/skill-use.ts
var SkillUseParseError = class extends Error {
	constructor(code, value) {
		super(formatSkillUseParseErrorMessage(code, value));
		this.name = "SkillUseParseError";
		this.code = code;
		this.value = value;
	}
};
function isSkillUseParseError(error) {
	return error instanceof SkillUseParseError;
}
function formatSkillUse(packageName, skillName) {
	const pkg = packageName.trim();
	const skill = skillName.trim();
	if (pkg === "") throw new SkillUseParseError("empty-package", `${packageName}#${skillName}`);
	if (skill === "") throw new SkillUseParseError("empty-skill", `${packageName}#${skillName}`);
	return `${pkg}#${skill}`;
}
function parseSkillUse(value) {
	const trimmed = value.trim();
	const separatorIndex = trimmed.indexOf("#");
	if (separatorIndex === -1) throw new SkillUseParseError("missing-separator", value);
	const packageName = trimmed.slice(0, separatorIndex).trim();
	const skillName = trimmed.slice(separatorIndex + 1).trim();
	if (packageName === "") throw new SkillUseParseError("empty-package", value);
	if (skillName === "") throw new SkillUseParseError("empty-skill", value);
	return {
		packageName,
		skillName
	};
}
function formatSkillUseParseErrorMessage(code, value) {
	switch (code) {
		case "missing-separator": return `Invalid skill use "${value}": expected <package>#<skill>.`;
		case "empty-package": return `Invalid skill use "${value}": package is required.`;
		case "empty-skill": return `Invalid skill use "${value}": skill is required.`;
	}
}

//#endregion
export { parseSkillUse as i, formatSkillUse as n, isSkillUseParseError as r, SkillUseParseError as t };
