// @tanstack/intent@0.0.44/dist/staleness-BGFfic-Y.mjs
import { c as parseFrontmatter, r as findSkillFiles, u as toPosixPath } from "./utils-9fhWAVua.mjs";
import { t as readIntentArtifacts } from "./artifact-coverage-DgWuVqUp.mjs";
import { existsSync, readFileSync } from "node:fs";
import { isAbsolute, join, relative, resolve } from "node:path";
import semver from "semver";

//#region src/staleness.ts
function classifyVersionDrift(oldVer, newVer) {
	const oldVersion = normalizeVersion(oldVer);
	const newVersion = normalizeVersion(newVer);
	if (!oldVersion || !newVersion) return null;
	if (semver.eq(oldVersion, newVersion)) return null;
	if (!semver.gt(newVersion, oldVersion)) return null;
	const oldParsed = semver.parse(oldVersion);
	const newParsed = semver.parse(newVersion);
	if (oldParsed && newParsed && oldParsed.major === newParsed.major && oldParsed.minor === newParsed.minor && oldParsed.patch === newParsed.patch && oldParsed.prerelease.length > 0) return "patch";
	switch (semver.diff(oldVersion, newVersion)) {
		case "major":
		case "premajor": return "major";
		case "minor":
		case "preminor": return "minor";
		case "patch":
		case "prepatch":
		case "prerelease": return "patch";
		default: return null;
	}
}
function normalizeVersion(version) {
	const validVersion = semver.valid(version);
	if (validVersion) return validVersion;
	return semver.coerce(version)?.version ?? null;
}
function readLocalVersion(packageDir) {
	try {
		const pkgJson = JSON.parse(readFileSync(join(packageDir, "package.json"), "utf8"));
		return typeof pkgJson.version === "string" ? pkgJson.version : null;
	} catch {
		return null;
	}
}
async function fetchNpmVersion(packageName) {
	try {
		const res = await fetch(`https://registry.npmjs.org/${encodeURIComponent(packageName)}/latest`);
		if (!res.ok) return null;
		const data = await res.json();
		return typeof data.version === "string" ? data.version : null;
	} catch {
		return null;
	}
}
async function fetchCurrentVersion(packageDir, packageName) {
	return readLocalVersion(packageDir) ?? await fetchNpmVersion(packageName);
}
function isStringRecord(value) {
	return !!value && typeof value === "object" && !Array.isArray(value) && Object.values(value).every((entry) => typeof entry === "string");
}
function parseSyncState(value) {
	if (!value || typeof value !== "object") return null;
	const raw = value;
	const parsed = {};
	if (typeof raw.library_version === "string") parsed.library_version = raw.library_version;
	if (raw.skills && typeof raw.skills === "object") {
		const skills = {};
		for (const [skillName, skillValue] of Object.entries(raw.skills)) {
			if (!skillValue || typeof skillValue !== "object") continue;
			const sourcesSha = skillValue.sources_sha;
			if (sourcesSha !== void 0 && !isStringRecord(sourcesSha)) continue;
			skills[skillName] = {};
			if (sourcesSha) skills[skillName].sources_sha = sourcesSha;
		}
		parsed.skills = skills;
	}
	return parsed;
}
function readSyncState(packageDir) {
	const statePath = join(packageDir, "skills", "sync-state.json");
	try {
		return parseSyncState(JSON.parse(readFileSync(statePath, "utf8")));
	} catch {
		return null;
	}
}
function readPackageName(packageDir) {
	const packageJson = readPackageJson(packageDir);
	return typeof packageJson?.name === "string" ? packageJson.name : relative(process.cwd(), packageDir) || "unknown";
}
function readPackageJson(packageDir) {
	try {
		return JSON.parse(readFileSync(join(packageDir, "package.json"), "utf8"));
	} catch {
		return null;
	}
}
function normalizeFilePath(path) {
	return toPosixPath(resolve(path));
}
function getRelativePackageDir(artifactRoot, packageDir) {
	return toPosixPath(relative(artifactRoot, packageDir));
}
function normalizeList(values) {
	return [...new Set(values ?? [])].sort((a, b) => a.localeCompare(b));
}
function sameStringList(a, b) {
	const left = normalizeList(a);
	const right = normalizeList(b);
	return left.length === right.length && left.every((value, index) => value === right[index]);
}
function artifactPackageMatches(artifact, packageDir, packageName, artifactRoot) {
	const relPackageDir = getRelativePackageDir(artifactRoot, packageDir);
	if (!relPackageDir) return true;
	if (artifact.packages.includes(packageName)) return true;
	if (artifact.packages.includes(relPackageDir)) return true;
	if (artifact.path?.startsWith(`${relPackageDir}/`)) return true;
	return artifact.packages.length === 0 && artifact.path === void 0;
}
function resolveArtifactSkillPaths(artifact, packageDir, artifactRoot) {
	if (!artifact.path) return [];
	const candidatePaths = [isAbsolute(artifact.path) ? artifact.path : join(artifactRoot, artifact.path), isAbsolute(artifact.path) ? artifact.path : join(packageDir, artifact.path)];
	if (artifact.package && artifact.path.startsWith("skills/")) candidatePaths.push(join(artifactRoot, artifact.package, artifact.path));
	return [...new Set(candidatePaths.map(normalizeFilePath))];
}
function findMatchingSkill(artifact, skillMetas, packageDir, artifactRoot) {
	const skillsByPath = new Map(skillMetas.map((skill) => [normalizeFilePath(skill.filePath), skill]));
	for (const candidatePath of resolveArtifactSkillPaths(artifact, packageDir, artifactRoot)) {
		const match = skillsByPath.get(candidatePath);
		if (match) return match;
	}
	const skillsByName = /* @__PURE__ */ new Map();
	for (const skill of skillMetas) {
		skillsByName.set(skill.name, skill);
		skillsByName.set(skill.relName, skill);
	}
	return (artifact.slug ? skillsByName.get(artifact.slug) : void 0) ?? (artifact.name ? skillsByName.get(artifact.name) : void 0) ?? null;
}
function buildArtifactSignals({ artifactRoot, artifacts, library, packageDir, skillMetas }) {
	if (!artifacts) return [];
	const artifactFiles = new Map([...artifacts.skillTrees, ...artifacts.domainMaps].map((file) => [file.path, file]));
	const signals = artifacts.warnings.map((warning) => ({
		type: "artifact-parse-warning",
		library,
		subject: warning.artifactPath,
		reasons: [warning.message],
		needsReview: true,
		artifactPath: warning.artifactPath
	}));
	for (const artifact of artifacts.skills) {
		if (!artifactPackageMatches(artifact, packageDir, library, artifactRoot)) continue;
		const subject = artifact.slug ?? artifact.name ?? artifact.path;
		const matchingSkill = findMatchingSkill(artifact, skillMetas, packageDir, artifactRoot);
		if (artifact.path && !matchingSkill) {
			signals.push({
				type: "artifact-skill-missing",
				library,
				subject,
				reasons: [`artifact skill path does not resolve to a generated SKILL.md (${artifact.path})`],
				needsReview: true,
				artifactPath: artifact.artifactPath,
				skill: artifact.slug ?? artifact.name
			});
			continue;
		}
		if (!matchingSkill) continue;
		if (matchingSkill.sources !== void 0 && artifact.sources.length > 0 && !sameStringList(matchingSkill.sources, artifact.sources)) signals.push({
			type: "artifact-source-drift",
			library,
			subject,
			reasons: ["artifact sources differ from SKILL.md frontmatter sources"],
			needsReview: true,
			artifactPath: artifact.artifactPath,
			skill: matchingSkill.name
		});
		const artifactVersion = artifactFiles.get(artifact.artifactPath)?.libraryVersion;
		if (artifactVersion && matchingSkill.libraryVersion && artifactVersion !== matchingSkill.libraryVersion) signals.push({
			type: "artifact-library-version-drift",
			library,
			subject,
			reasons: [`artifact library.version (${artifactVersion}) differs from SKILL.md library_version (${matchingSkill.libraryVersion})`],
			needsReview: true,
			artifactPath: artifact.artifactPath,
			skill: matchingSkill.name
		});
	}
	return signals;
}
function artifactCoversPackage(artifact, packageDir, packageName, artifactRoot) {
	const relPackageDir = getRelativePackageDir(artifactRoot, packageDir);
	return artifact.packages.includes(packageName) || artifact.packages.includes(relPackageDir) || artifact.package === packageName || artifact.package === relPackageDir || artifact.path?.startsWith(`${relPackageDir}/`) === true;
}
function artifactIgnoresPackage(artifacts, packageDir, packageName, artifactRoot) {
	const relPackageDir = getRelativePackageDir(artifactRoot, packageDir);
	return artifacts.ignoredPackages.some((ignored) => ignored.packageName === packageName || ignored.packageName === relPackageDir);
}
function buildWorkspaceCoverageSignals({ artifactRoot, artifacts, packageDirs }) {
	if (!artifacts) return [];
	const signals = [];
	for (const packageDir of packageDirs) {
		if (readPackageJson(packageDir)?.private === true) continue;
		const packageName = readPackageName(packageDir);
		if (artifactIgnoresPackage(artifacts, packageDir, packageName, artifactRoot)) continue;
		const hasGeneratedSkill = findSkillFiles(join(packageDir, "skills")).length > 0;
		const hasArtifactCoverage = artifacts.skills.some((artifact) => artifactCoversPackage(artifact, packageDir, packageName, artifactRoot));
		if (hasGeneratedSkill || hasArtifactCoverage) continue;
		signals.push({
			type: "missing-package-coverage",
			library: packageName,
			subject: packageName,
			reasons: ["workspace package is not represented by generated skills or _artifacts coverage"],
			needsReview: true,
			packageName,
			packageRoot: getRelativePackageDir(artifactRoot, packageDir)
		});
	}
	return signals;
}
async function checkStaleness(packageDir, packageName, artifactRoot = packageDir) {
	const skillsDir = join(packageDir, "skills");
	const library = packageName ?? readPackageName(packageDir);
	const skillMetas = findSkillFiles(skillsDir).map((filePath) => {
		const fm = parseFrontmatter(filePath);
		const relName = toPosixPath(relative(skillsDir, filePath)).replace(/[/\\]SKILL\.md$/, "");
		return {
			name: typeof fm?.name === "string" ? fm.name : relName,
			relName,
			filePath,
			libraryVersion: fm?.library_version,
			sources: Array.isArray(fm?.sources) ? fm.sources : void 0
		};
	});
	const artifacts = existsSync(join(artifactRoot, "_artifacts")) ? readIntentArtifacts(artifactRoot) : null;
	const skillVersion = skillMetas.find((s) => s.libraryVersion)?.libraryVersion ?? null;
	const currentVersion = await fetchCurrentVersion(packageDir, library);
	const versionDrift = skillVersion && currentVersion ? classifyVersionDrift(skillVersion, currentVersion) : null;
	const syncState = readSyncState(packageDir);
	return {
		library,
		currentVersion,
		skillVersion,
		versionDrift,
		skills: skillMetas.map((skill) => {
			const reasons = [];
			if (currentVersion && skill.libraryVersion && classifyVersionDrift(skill.libraryVersion, currentVersion) !== null) reasons.push(`version drift (${skill.libraryVersion} → ${currentVersion})`);
			const storedShas = syncState?.skills?.[skill.name]?.sources_sha ?? {};
			if (skill.sources && Object.keys(storedShas).length > 0) {
				for (const source of skill.sources) if (!storedShas[source]) reasons.push(`new source (${source})`);
			}
			return {
				name: skill.name,
				reasons,
				needsReview: reasons.length > 0
			};
		}),
		signals: buildArtifactSignals({
			artifactRoot,
			artifacts,
			library,
			packageDir,
			skillMetas
		})
	};
}

//#endregion
export { checkStaleness as n, readPackageName as r, buildWorkspaceCoverageSignals as t };
