// @tanstack/intent@0.0.44/dist/utils-9fhWAVua.mjs
import { createRequire } from "node:module";
import { closeSync, existsSync, lstatSync, openSync, readFileSync, readSync, readdirSync, realpathSync } from "node:fs";
import { dirname, join, resolve, sep } from "node:path";
import { execFileSync } from "node:child_process";
import { parse } from "yaml";

//#region src/utils.ts
const nodeReadFs = {
	existsSync,
	lstatSync,
	readFileSync,
	readdirSync,
	realpathSync,
	openSync,
	readSync,
	closeSync
};
/**
* Convert a path to use forward slashes (for cross-platform consistency).
*/
function toPosixPath(p) {
	return p.split(sep).join("/");
}
function createFsIdentityCache(getFs = () => nodeReadFs) {
	const cache = /* @__PURE__ */ new Map();
	return (path) => {
		const resolved = resolve(path);
		const cached = cache.get(resolved);
		if (cached) return cached;
		const fs = getFs();
		let identity;
		try {
			identity = fs.lstatSync(resolved).isSymbolicLink() ? fs.realpathSync(resolved) : resolved;
		} catch {
			identity = resolved;
		}
		cache.set(resolved, identity);
		return identity;
	};
}
/**
* Recursively find all SKILL.md files under a directory.
*/
function findSkillFiles(dir, fs = nodeReadFs) {
	const files = [];
	collectSkillFiles(dir, fs, files);
	return files;
}
function collectSkillFiles(dir, fs, files) {
	let entries;
	try {
		entries = fs.readdirSync(dir, {
			withFileTypes: true,
			encoding: "utf8"
		});
	} catch {
		return;
	}
	for (const entry of entries) {
		const fullPath = join(dir, entry.name);
		if (entry.isDirectory()) collectSkillFiles(fullPath, fs, files);
		else if (entry.name === "SKILL.md") files.push(fullPath);
	}
}
/**
* Read dependencies and peerDependencies (and optionally devDependencies) from
* a parsed package.json object.
*/
function getDeps(pkgJson, includeDevDeps = false) {
	const deps = /* @__PURE__ */ new Set();
	const fields = includeDevDeps ? [
		"dependencies",
		"devDependencies",
		"peerDependencies"
	] : ["dependencies", "peerDependencies"];
	for (const field of fields) {
		const d = pkgJson[field];
		if (d && typeof d === "object") for (const name of Object.keys(d)) deps.add(name);
	}
	return [...deps];
}
function listNodeModulesPackageDirs(nodeModulesDir) {
	let topEntries;
	try {
		topEntries = readdirSync(nodeModulesDir, {
			withFileTypes: true,
			encoding: "utf8"
		});
	} catch {
		return [];
	}
	const packageDirs = [];
	for (const entry of topEntries) {
		if (!entry.isDirectory() && !entry.isSymbolicLink()) continue;
		const dirPath = join(nodeModulesDir, entry.name);
		if (entry.name.startsWith("@")) {
			let scopedEntries;
			try {
				scopedEntries = readdirSync(dirPath, {
					withFileTypes: true,
					encoding: "utf8"
				});
			} catch {
				continue;
			}
			for (const scoped of scopedEntries) {
				if (!scoped.isDirectory() && !scoped.isSymbolicLink()) continue;
				packageDirs.push(join(dirPath, scoped.name));
			}
		} else if (!entry.name.startsWith(".")) packageDirs.push(dirPath);
	}
	return packageDirs;
}
function listNestedNodeModulesPackageDirs(nodeModulesDir, getFsIdentity = createFsIdentityCache()) {
	const packageDirs = [];
	const visitedNodeModulesDirs = /* @__PURE__ */ new Set();
	const visitedPackageDirs = /* @__PURE__ */ new Set();
	function readDir(dir) {
		try {
			return readdirSync(dir, {
				withFileTypes: true,
				encoding: "utf8"
			});
		} catch {
			return [];
		}
	}
	function addPackageDir(packageDir) {
		const key = getFsIdentity(packageDir);
		if (visitedPackageDirs.has(key)) return;
		visitedPackageDirs.add(key);
		if (existsSync(join(packageDir, "package.json"))) packageDirs.push(packageDir);
		scanNodeModulesDir(join(packageDir, "node_modules"));
	}
	function scanNodeModulesDir(dir) {
		const key = getFsIdentity(dir);
		if (visitedNodeModulesDirs.has(key)) return;
		visitedNodeModulesDirs.add(key);
		for (const entry of readDir(dir)) {
			if (!entry.isDirectory() && !entry.isSymbolicLink()) continue;
			const dirPath = join(dir, entry.name);
			if (entry.name.startsWith("@")) {
				for (const scoped of readDir(dirPath)) {
					if (!scoped.isDirectory() && !scoped.isSymbolicLink()) continue;
					addPackageDir(join(dirPath, scoped.name));
				}
				continue;
			}
			if (!entry.name.startsWith(".")) addPackageDir(dirPath);
		}
	}
	scanNodeModulesDir(nodeModulesDir);
	return packageDirs;
}
function detectGlobalNodeModules(packageManager) {
	const envPath = process.env.INTENT_GLOBAL_NODE_MODULES?.trim();
	if (envPath) return {
		path: envPath,
		source: "INTENT_GLOBAL_NODE_MODULES"
	};
	const commands = [];
	if (packageManager === "pnpm") commands.push({
		command: "pnpm",
		args: ["root", "-g"]
	});
	if (packageManager === "yarn") commands.push({
		command: "yarn",
		args: ["global", "dir"],
		transform: (output) => join(output, "node_modules")
	});
	commands.push({
		command: "npm",
		args: ["root", "-g"]
	});
	for (const candidate of commands) try {
		const output = execFileSync(candidate.command, candidate.args, {
			encoding: "utf8",
			stdio: [
				"ignore",
				"pipe",
				"ignore"
			]
		}).trim();
		if (!output) continue;
		return {
			path: candidate.transform ? candidate.transform(output) : output,
			source: `${candidate.command} ${candidate.args.join(" ")}`
		};
	} catch {
		continue;
	}
	return { path: null };
}
/**
* Resolve the directory of a dependency by name. Tries createRequire first
* (handles pnpm symlinks), then falls back to walking up node_modules
* directories (handles packages with export maps that block ./package.json).
*/
/**
* `createRequire` builds a full module-resolution context; constructing it is
* non-trivial and `resolveDepDir` is called once per dependency, often many
* times from the same `parentDir` (every sibling dep of one package). Cache the
* require function by its base `package.json` path. `req.resolve` still hits the
* live filesystem on each call, so cached entries never go stale.
*/
const requireForBaseCache = /* @__PURE__ */ new Map();
function getRequireForBase(basePackageJson) {
	let req = requireForBaseCache.get(basePackageJson);
	if (!req) {
		req = createRequire(basePackageJson);
		requireForBaseCache.set(basePackageJson, req);
	}
	return req;
}
function resolveDepDir(depName, parentDir) {
	try {
		return dirname(getRequireForBase(join(parentDir, "package.json")).resolve(join(depName, "package.json")));
	} catch (err) {
		const code = err && typeof err === "object" && "code" in err ? err.code : void 0;
		if (code && code !== "MODULE_NOT_FOUND" && code !== "ERR_PACKAGE_PATH_NOT_EXPORTED") console.warn(`Warning: could not resolve ${depName} from ${parentDir}: ${err instanceof Error ? err.message : String(err)}`);
	}
	let dir = parentDir;
	let prev;
	while (dir !== prev) {
		const candidate = join(dir, "node_modules", depName);
		if (existsSync(join(candidate, "package.json"))) return candidate;
		prev = dir;
		dir = dirname(dir);
	}
	return null;
}
/**
* Parse YAML frontmatter from a file. Returns null if no frontmatter or on error.
*/
function parseFrontmatter(filePath, fs = nodeReadFs) {
	const content = readFrontmatterRegion(filePath, fs);
	if (content === null) return null;
	const match = content.match(/^---\r?\n([\s\S]*?)\r?\n---/);
	if (!match?.[1]) return null;
	try {
		return parse(match[1]);
	} catch {
		return null;
	}
}
/** Max bytes read when probing a file's leading frontmatter region. */
const FRONTMATTER_READ_LIMIT = 16 * 1024;
/** Reused across calls; safe because reads are synchronous and single-threaded. */
const frontmatterBuffer = Buffer.allocUnsafe(FRONTMATTER_READ_LIMIT);
/**
* Read just the leading region of a file, enough to cover its frontmatter,
* instead of its whole body. Falls back to a full read when the bounded read
* primitives are unavailable or the frontmatter exceeds the probe limit.
*/
function readFrontmatterRegion(filePath, fs) {
	if (fs.openSync && fs.readSync && fs.closeSync) {
		let region = null;
		let truncated = false;
		let fd;
		try {
			fd = fs.openSync(filePath, "r");
		} catch {
			return null;
		}
		try {
			const bytesRead = fs.readSync(fd, frontmatterBuffer, 0, FRONTMATTER_READ_LIMIT, 0);
			region = frontmatterBuffer.toString("utf8", 0, bytesRead);
			truncated = bytesRead === FRONTMATTER_READ_LIMIT && !/\r?\n---/.test(region.slice(3));
		} finally {
			fs.closeSync(fd);
		}
		if (!truncated) return region;
	}
	try {
		return fs.readFileSync(filePath, "utf8");
	} catch {
		return null;
	}
}

//#endregion
export { listNestedNodeModulesPackageDirs as a, parseFrontmatter as c, getDeps as i, resolveDepDir as l, detectGlobalNodeModules as n, listNodeModulesPackageDirs as o, findSkillFiles as r, nodeReadFs as s, createFsIdentityCache as t, toPosixPath as u };
