// @tanstack/intent@0.0.44/dist/workspace-patterns-BffPlZ1D.mjs
import { r as findSkillFiles } from "./utils-9fhWAVua.mjs";
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { parse } from "yaml";
import { parse as parse$1 } from "jsonc-parser";

//#region src/workspace-patterns.ts
function normalizeWorkspacePattern(pattern) {
	return pattern.replace(/\\/g, "/").replace(/^\.\//, "").replace(/\/+$/, "");
}
function normalizeWorkspacePatterns(patterns) {
	return [...new Set(patterns.map(normalizeWorkspacePattern).filter(Boolean))].sort((a, b) => a.localeCompare(b));
}
function isRecord(value) {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}
function parseWorkspacePatternList(value, fieldName) {
	if (value === void 0 || value === null) return null;
	if (!Array.isArray(value)) throw new TypeError(`${fieldName} must be an array of strings`);
	if (value.some((pattern) => typeof pattern !== "string")) throw new TypeError(`${fieldName} must be an array of strings`);
	return normalizeWorkspacePatterns(value);
}
function parseWorkspacePatternField(value, fieldName, nestedKey) {
	if (value === void 0 || value === null) return null;
	if (Array.isArray(value)) return parseWorkspacePatternList(value, fieldName);
	if (isRecord(value)) return parseWorkspacePatternList(value[nestedKey], `${fieldName}.${nestedKey}`);
	throw new TypeError(`${fieldName} must be an array of strings or an object with ${nestedKey}`);
}
function hasWorkspaceManifest(dir) {
	return existsSync(join(dir, "package.json")) || existsSync(join(dir, "deno.json")) || existsSync(join(dir, "deno.jsonc"));
}
function readYamlFile(path) {
	return parse(readFileSync(path, "utf8"));
}
function readJsonFile(path) {
	return JSON.parse(readFileSync(path, "utf8"));
}
function readJsoncFile(path) {
	const errors = [];
	const value = parse$1(readFileSync(path, "utf8"), errors, { allowTrailingComma: true });
	if (errors.length > 0) throw new SyntaxError(errors.map((error) => `JSONC parse error ${error.error} at offset ${error.offset}`).join("; "));
	return value;
}
function warnConfigError(path, err) {
	const verb = err instanceof SyntaxError ? "parse" : "read";
	console.error(`Warning: failed to ${verb} ${path}: ${err instanceof Error ? err.message : err}`);
}
const workspacePatternSources = [
	{
		fileName: "pnpm-workspace.yaml",
		read: readYamlFile,
		getPatterns: (config) => parseWorkspacePatternList(isRecord(config) ? config.packages : void 0, "pnpm-workspace.yaml#packages")
	},
	{
		fileName: "package.json",
		read: readJsonFile,
		getPatterns: (config) => parseWorkspacePatternField(isRecord(config) ? config.workspaces : void 0, "package.json#workspaces", "packages")
	},
	{
		fileName: "deno.json",
		read: readJsoncFile,
		getPatterns: (config) => parseWorkspacePatternField(isRecord(config) ? config.workspace : void 0, "deno.json#workspace", "members")
	},
	{
		fileName: "deno.jsonc",
		read: readJsoncFile,
		getPatterns: (config) => parseWorkspacePatternField(isRecord(config) ? config.workspace : void 0, "deno.jsonc#workspace", "members")
	}
];
const workspacePatternsCache = /* @__PURE__ */ new Map();
const workspaceRootCache = /* @__PURE__ */ new Map();
const workspacePackageDirsCache = /* @__PURE__ */ new Map();
const workspaceInfoCache = /* @__PURE__ */ new Map();
function readWorkspacePatterns(root) {
	if (workspacePatternsCache.has(root)) return workspacePatternsCache.get(root) ?? null;
	const patterns = readWorkspacePatternsUncached(root);
	workspacePatternsCache.set(root, patterns);
	return patterns;
}
function readWorkspacePatternsUncached(root) {
	for (const source of workspacePatternSources) {
		const path = join(root, source.fileName);
		if (!existsSync(path)) continue;
		try {
			const patterns = source.getPatterns(source.read(path));
			if (patterns) return patterns;
		} catch (err) {
			warnConfigError(path, err);
		}
	}
	return null;
}
function readWorkspacePackageDirs(root) {
	if (workspacePackageDirsCache.has(root)) return workspacePackageDirsCache.get(root) ?? null;
	const patterns = readWorkspacePatterns(root);
	if (!patterns) {
		workspacePackageDirsCache.set(root, null);
		return null;
	}
	const packageDirs = resolveWorkspacePackages(root, patterns);
	workspacePackageDirsCache.set(root, packageDirs);
	return packageDirs;
}
function getWorkspaceInfo(root) {
	if (workspaceInfoCache.has(root)) return workspaceInfoCache.get(root) ?? null;
	const patterns = readWorkspacePatterns(root);
	if (!patterns) {
		workspaceInfoCache.set(root, null);
		return null;
	}
	const packageDirs = readWorkspacePackageDirs(root) ?? [];
	const info = {
		root,
		patterns,
		packageDirs,
		packageDirsWithSkills: packageDirs.filter((dir) => {
			const skillsDir = join(dir, "skills");
			return existsSync(skillsDir) && findSkillFiles(skillsDir).length > 0;
		})
	};
	workspaceInfoCache.set(root, info);
	return info;
}
function resolveWorkspacePackages(root, patterns) {
	const includedDirs = /* @__PURE__ */ new Set();
	const excludedDirs = /* @__PURE__ */ new Set();
	for (const pattern of normalizeWorkspacePatterns(patterns)) {
		if (pattern.startsWith("!")) {
			resolveWorkspacePatternSegments(root, pattern.slice(1).split("/"), excludedDirs);
			continue;
		}
		resolveWorkspacePatternSegments(root, pattern.split("/"), includedDirs);
	}
	return [...includedDirs].filter((dir) => !excludedDirs.has(dir)).sort((a, b) => a.localeCompare(b));
}
/** Recursively matches path segments: `*` matches one level, `**` matches zero or more levels. */
function resolveWorkspacePatternSegments(dir, segments, result) {
	if (segments.length === 0) {
		if (hasWorkspaceManifest(dir)) result.add(dir);
		return;
	}
	const segment = segments[0];
	const remainingSegments = segments.slice(1);
	if (segment === "**") {
		resolveWorkspacePatternSegments(dir, remainingSegments, result);
		for (const childDir of readChildDirectories(dir)) resolveWorkspacePatternSegments(childDir, segments, result);
		return;
	}
	if (segment === "*") {
		for (const childDir of readChildDirectories(dir)) resolveWorkspacePatternSegments(childDir, remainingSegments, result);
		return;
	}
	const nextDir = join(dir, segment);
	if (!existsSync(nextDir)) return;
	resolveWorkspacePatternSegments(nextDir, remainingSegments, result);
}
function readChildDirectories(dir) {
	try {
		return readdirSync(dir, { withFileTypes: true }).filter((entry) => entry.isDirectory() && entry.name !== "node_modules" && !entry.name.startsWith(".")).map((entry) => join(dir, entry.name));
	} catch (err) {
		console.error(`Warning: could not read directory ${dir}: ${err instanceof Error ? err.message : err}`);
		return [];
	}
}
function findWorkspaceRoot(start) {
	let dir = start;
	let prev;
	const visited = [];
	while (dir !== prev) {
		const cached = workspaceRootCache.get(dir);
		if (cached !== void 0) {
			for (const visitedDir of visited) workspaceRootCache.set(visitedDir, cached);
			return cached;
		}
		visited.push(dir);
		if (readWorkspacePatterns(dir)) {
			for (const visitedDir of visited) workspaceRootCache.set(visitedDir, dir);
			return dir;
		}
		prev = dir;
		dir = dirname(dir);
	}
	for (const visitedDir of visited) workspaceRootCache.set(visitedDir, null);
	return null;
}
function findPackagesWithSkills(root) {
	return getWorkspaceInfo(root)?.packageDirsWithSkills ?? [];
}
function findWorkspacePackages(root) {
	return readWorkspacePackageDirs(root) ?? [];
}

//#endregion
export { readWorkspacePatterns as a, getWorkspaceInfo as i, findWorkspacePackages as n, resolveWorkspacePackages as o, findWorkspaceRoot as r, findPackagesWithSkills as t };
